{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-Assembling"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "### --- Libraries importieren ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import re\n",
    "import csv\n",
    "\n",
    "from scipy.sparse import coo_matrix\n",
    "from scipy.io import mmwrite\n",
    "from pathlib import Path\n",
    "\n",
    "import sctoolbox\n",
    "import sctoolbox.utils as utils\n",
    "import sctoolbox.utils.assemblers as assembler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### --- Datei einlesen ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/stud9/notebooks/rna_analysis/notebooks/ScRNA-analysis\n"
     ]
    }
   ],
   "source": [
    "file_path_pat = \"/mnt/workspace_stud/napkon_data/wp2_rna/out.txt\"\n",
    "file_path_gene = \"/mnt/workspace_stud/napkon_data/wp2_rna/combined_rna_napkon_4_all.matrix.raw.anno\"\n",
    "\n",
    "\n",
    "path_mtx = str(Path.cwd()) # Aktueller Speicherort/Verzeichnis mit .mtx, barcodes.tsv und genes.tsv\n",
    "print(path_mtx)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. --- Patienten-Metadaten lesen und vorbereiten für Assemblierung ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TubeID</th>\n",
       "      <th>meta TubeID</th>\n",
       "      <th>meta pop export_psn</th>\n",
       "      <th>meta pop visit_1_date</th>\n",
       "      <th>meta pop visit_2_date</th>\n",
       "      <th>meta pop visit_3_date</th>\n",
       "      <th>meta pop visit_4_date</th>\n",
       "      <th>meta pop age</th>\n",
       "      <th>meta pop age_dec</th>\n",
       "      <th>meta pop sex</th>\n",
       "      <th>...</th>\n",
       "      <th>rna unassigned reads due to bad mapping quality</th>\n",
       "      <th>rna unassigned reads due to low fragment length</th>\n",
       "      <th>rna unassigned chimeric reads</th>\n",
       "      <th>rna usable sample: mapped reads &gt; 2000000</th>\n",
       "      <th>rna project.run id</th>\n",
       "      <th>rna DDX3Y</th>\n",
       "      <th>rna XIST</th>\n",
       "      <th>rna gender</th>\n",
       "      <th>rna sample id original</th>\n",
       "      <th>rna plate id original</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3771</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>plt33.run1 pat_2384</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sample</td>\n",
       "      <td>Plate 33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3772</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>plt38.run1 pat_2741</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sample 1</td>\n",
       "      <td>Plate 38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3773</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>plt40.run1 pat_3006</td>\n",
       "      <td>809.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Sample 27</td>\n",
       "      <td>Plate 40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3774</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>plt40.run1 pat_3033</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sample 32</td>\n",
       "      <td>Plate 40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3775</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>plt1.run1 pat_146</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>Sample 1</td>\n",
       "      <td>Plate 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     TubeID meta TubeID meta pop export_psn meta pop visit_1_date  \\\n",
       "3771      0           0                   0                     0   \n",
       "3772      0           0                   0                     0   \n",
       "3773      0           0                   0                     0   \n",
       "3774      0           0                   0                     0   \n",
       "3775      0           0                   0                     0   \n",
       "\n",
       "     meta pop visit_2_date meta pop visit_3_date meta pop visit_4_date  \\\n",
       "3771                     0                     0                     0   \n",
       "3772                     0                     0                     0   \n",
       "3773                     0                     0                     0   \n",
       "3774                     0                     0                     0   \n",
       "3775                     0                     0                     0   \n",
       "\n",
       "      meta pop age meta pop age_dec meta pop sex  ...  \\\n",
       "3771           0.0                0            0  ...   \n",
       "3772           0.0                0            0  ...   \n",
       "3773           0.0                0            0  ...   \n",
       "3774           0.0                0            0  ...   \n",
       "3775           0.0                0            0  ...   \n",
       "\n",
       "     rna unassigned reads due to bad mapping quality  \\\n",
       "3771                                             0.0   \n",
       "3772                                             0.0   \n",
       "3773                                             0.0   \n",
       "3774                                             0.0   \n",
       "3775                                             0.0   \n",
       "\n",
       "     rna unassigned reads due to low fragment length  \\\n",
       "3771                                             0.0   \n",
       "3772                                             0.0   \n",
       "3773                                             0.0   \n",
       "3774                                             0.0   \n",
       "3775                                             0.0   \n",
       "\n",
       "     rna unassigned chimeric reads rna usable sample: mapped reads > 2000000  \\\n",
       "3771                           0.0                                     False   \n",
       "3772                           0.0                                     False   \n",
       "3773                           0.0                                      True   \n",
       "3774                           0.0                                     False   \n",
       "3775                           0.0                                     False   \n",
       "\n",
       "       rna project.run id rna DDX3Y rna XIST rna gender  \\\n",
       "3771  plt33.run1 pat_2384       0.0      0.0          0   \n",
       "3772  plt38.run1 pat_2741       0.0      0.0          0   \n",
       "3773  plt40.run1 pat_3006     809.0      4.0          M   \n",
       "3774  plt40.run1 pat_3033       0.0      0.0          0   \n",
       "3775    plt1.run1 pat_146       0.0      0.0          0   \n",
       "\n",
       "     rna sample id original rna plate id original  \n",
       "3771                Sample               Plate 33  \n",
       "3772               Sample 1              Plate 38  \n",
       "3773              Sample 27              Plate 40  \n",
       "3774              Sample 32              Plate 40  \n",
       "3775               Sample 1               Plate 1  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pat_df = pd.read_csv(file_path_pat, decimal=\",\", sep=\"\\t\")\n",
    "pat_df = pat_df.fillna(0)\n",
    "pat_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                         0\n",
       "1          plt5.run1 pat_26\n",
       "2                         0\n",
       "3                         0\n",
       "4         plt2.run1 pat_195\n",
       "               ...         \n",
       "3771    plt33.run1 pat_2384\n",
       "3772    plt38.run1 pat_2741\n",
       "3773    plt40.run1 pat_3006\n",
       "3774    plt40.run1 pat_3033\n",
       "3775      plt1.run1 pat_146\n",
       "Name: rna project.run id, Length: 3776, dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pat_df[\"rna project.run id\"] # ID überprüfen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# optional: Patienten-Datei im lokalen Speicherort abspeichern\n",
    "# pat_df.to_csv('pat_meta.csv', index=False) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. --- Gen-Metadaten lesen und vorbereiten für Assemblierung ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl gene id</th>\n",
       "      <th>Ensembl gene</th>\n",
       "      <th>Ensembl biotype</th>\n",
       "      <th>UniProt proteins</th>\n",
       "      <th>UniProt genes</th>\n",
       "      <th>UniProt accessions</th>\n",
       "      <th>UniProt names</th>\n",
       "      <th>UniProt Ensembl transcripts</th>\n",
       "      <th>UniProt Ensembl proteins</th>\n",
       "      <th>UniProt Ensembl gene ids</th>\n",
       "      <th>...</th>\n",
       "      <th>plt49.run1 pat_3563</th>\n",
       "      <th>plt49.run1 pat_3564</th>\n",
       "      <th>plt49.run1 pat_3565</th>\n",
       "      <th>plt49.run1 pat_3566</th>\n",
       "      <th>plt49.run1 pat_3567</th>\n",
       "      <th>plt49.run1 pat_3568</th>\n",
       "      <th>plt49.run1 pat_3569</th>\n",
       "      <th>plt49.run1 pat_3570</th>\n",
       "      <th>plt49.run1 pat_3571</th>\n",
       "      <th>plt49.run1 pat_3572</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000146555</td>\n",
       "      <td>SDK1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Protein sidekick-1</td>\n",
       "      <td>SDK1</td>\n",
       "      <td>Q7Z5N4; F8W6X9; A0A087WTQ6</td>\n",
       "      <td>SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN</td>\n",
       "      <td>ENST00000404826; ENST00000389531; ENST00000615806</td>\n",
       "      <td>ENSP00000385899; ENSP00000374182; ENSP00000478062</td>\n",
       "      <td>ENSG00000146555</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000166523</td>\n",
       "      <td>CLEC4E</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C-type lectin domain family 4 member E</td>\n",
       "      <td>CLEC4E, CLECSF9, MINCLE, UNQ218/PRO244; CLEC4E</td>\n",
       "      <td>Q9ULY5; F8WFA1; F5H5X7</td>\n",
       "      <td>CLC4E_HUMAN; F8WFA1_HUMAN; F5H5X7_HUMAN</td>\n",
       "      <td>ENST00000299663; ENST00000446457, ENST00000450...</td>\n",
       "      <td>ENSP00000299663; ENSP00000387737, ENSP00000404...</td>\n",
       "      <td>ENSG00000166523</td>\n",
       "      <td>...</td>\n",
       "      <td>198</td>\n",
       "      <td>96</td>\n",
       "      <td>90</td>\n",
       "      <td>287</td>\n",
       "      <td>309</td>\n",
       "      <td>94</td>\n",
       "      <td>62</td>\n",
       "      <td>215</td>\n",
       "      <td>261</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000125945</td>\n",
       "      <td>ZNF436</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Zinc finger protein 436</td>\n",
       "      <td>ZNF436, KIAA1710</td>\n",
       "      <td>Q9C0F3</td>\n",
       "      <td>ZN436_HUMAN</td>\n",
       "      <td>ENST00000314011, ENST00000374608, ENST00000635...</td>\n",
       "      <td>ENSP00000313582, ENSP00000363736, ENSP00000489...</td>\n",
       "      <td>ENSG00000125945</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>21</td>\n",
       "      <td>70</td>\n",
       "      <td>39</td>\n",
       "      <td>71</td>\n",
       "      <td>15</td>\n",
       "      <td>26</td>\n",
       "      <td>66</td>\n",
       "      <td>41</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000206145</td>\n",
       "      <td>P2RX6P</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSG00000206145</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3679 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Ensembl gene id     Ensembl gene                     Ensembl biotype  \\\n",
       "0  ENSG00000146555             SDK1                      protein_coding   \n",
       "1  ENSG00000283537  ENSG00000283537  transcribed_unprocessed_pseudogene   \n",
       "2  ENSG00000166523           CLEC4E                      protein_coding   \n",
       "3  ENSG00000125945           ZNF436                      protein_coding   \n",
       "4  ENSG00000206145           P2RX6P  transcribed_unprocessed_pseudogene   \n",
       "\n",
       "                         UniProt proteins  \\\n",
       "0                      Protein sidekick-1   \n",
       "1                                           \n",
       "2  C-type lectin domain family 4 member E   \n",
       "3                 Zinc finger protein 436   \n",
       "4                                           \n",
       "\n",
       "                                    UniProt genes          UniProt accessions  \\\n",
       "0                                            SDK1  Q7Z5N4; F8W6X9; A0A087WTQ6   \n",
       "1                                                                               \n",
       "2  CLEC4E, CLECSF9, MINCLE, UNQ218/PRO244; CLEC4E      Q9ULY5; F8WFA1; F5H5X7   \n",
       "3                                ZNF436, KIAA1710                      Q9C0F3   \n",
       "4                                                                               \n",
       "\n",
       "                                UniProt names  \\\n",
       "0  SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN   \n",
       "1                                               \n",
       "2     CLC4E_HUMAN; F8WFA1_HUMAN; F5H5X7_HUMAN   \n",
       "3                                 ZN436_HUMAN   \n",
       "4                                               \n",
       "\n",
       "                         UniProt Ensembl transcripts  \\\n",
       "0  ENST00000404826; ENST00000389531; ENST00000615806   \n",
       "1                                                      \n",
       "2  ENST00000299663; ENST00000446457, ENST00000450...   \n",
       "3  ENST00000314011, ENST00000374608, ENST00000635...   \n",
       "4                                                      \n",
       "\n",
       "                            UniProt Ensembl proteins UniProt Ensembl gene ids  \\\n",
       "0  ENSP00000385899; ENSP00000374182; ENSP00000478062          ENSG00000146555   \n",
       "1                                                             ENSG00000283537   \n",
       "2  ENSP00000299663; ENSP00000387737, ENSP00000404...          ENSG00000166523   \n",
       "3  ENSP00000313582, ENSP00000363736, ENSP00000489...          ENSG00000125945   \n",
       "4                                                             ENSG00000206145   \n",
       "\n",
       "   ... plt49.run1 pat_3563 plt49.run1 pat_3564 plt49.run1 pat_3565  \\\n",
       "0  ...                   5                   2                   2   \n",
       "1  ...                   0                   0                   1   \n",
       "2  ...                 198                  96                  90   \n",
       "3  ...                  40                  21                  70   \n",
       "4  ...                   0                   0                   0   \n",
       "\n",
       "  plt49.run1 pat_3566 plt49.run1 pat_3567 plt49.run1 pat_3568  \\\n",
       "0                   0                   2                   0   \n",
       "1                   0                   0                   0   \n",
       "2                 287                 309                  94   \n",
       "3                  39                  71                  15   \n",
       "4                   0                   0                   1   \n",
       "\n",
       "  plt49.run1 pat_3569 plt49.run1 pat_3570 plt49.run1 pat_3571  \\\n",
       "0                   1                   2                   1   \n",
       "1                   0                   0                   0   \n",
       "2                  62                 215                 261   \n",
       "3                  26                  66                  41   \n",
       "4                   0                   1                   0   \n",
       "\n",
       "  plt49.run1 pat_3572  \n",
       "0                   0  \n",
       "1                   0  \n",
       "2                 131  \n",
       "3                  17  \n",
       "4                   0  \n",
       "\n",
       "[5 rows x 3679 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = []\n",
    "\n",
    "with open(file_path_gene, 'r') as file:\n",
    "    content = file.read()\n",
    "    lines = content.splitlines()\n",
    "\n",
    "    header = lines[0].split('\\t')\n",
    "\n",
    "    for line in lines[1:]:\n",
    "        elements = line.split('\\t')\n",
    "        data.append(elements)\n",
    "\n",
    "gene_df = pd.DataFrame(data, columns=header)\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl gene id</th>\n",
       "      <th>Ensembl gene</th>\n",
       "      <th>Ensembl biotype</th>\n",
       "      <th>UniProt proteins</th>\n",
       "      <th>UniProt genes</th>\n",
       "      <th>UniProt accessions</th>\n",
       "      <th>UniProt names</th>\n",
       "      <th>UniProt Ensembl transcripts</th>\n",
       "      <th>UniProt Ensembl proteins</th>\n",
       "      <th>UniProt Ensembl gene ids</th>\n",
       "      <th>...</th>\n",
       "      <th>plt49.run1 pat_3563</th>\n",
       "      <th>plt49.run1 pat_3564</th>\n",
       "      <th>plt49.run1 pat_3565</th>\n",
       "      <th>plt49.run1 pat_3566</th>\n",
       "      <th>plt49.run1 pat_3567</th>\n",
       "      <th>plt49.run1 pat_3568</th>\n",
       "      <th>plt49.run1 pat_3569</th>\n",
       "      <th>plt49.run1 pat_3570</th>\n",
       "      <th>plt49.run1 pat_3571</th>\n",
       "      <th>plt49.run1 pat_3572</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000146555</td>\n",
       "      <td>SDK1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Protein sidekick-1</td>\n",
       "      <td>SDK1</td>\n",
       "      <td>Q7Z5N4; F8W6X9; A0A087WTQ6</td>\n",
       "      <td>SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN</td>\n",
       "      <td>ENST00000404826; ENST00000389531; ENST00000615806</td>\n",
       "      <td>ENSP00000385899; ENSP00000374182; ENSP00000478062</td>\n",
       "      <td>ENSG00000146555</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000166523</td>\n",
       "      <td>CLEC4E</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C-type lectin domain family 4 member E</td>\n",
       "      <td>CLEC4E, CLECSF9, MINCLE, UNQ218/PRO244; CLEC4E</td>\n",
       "      <td>Q9ULY5; F8WFA1; F5H5X7</td>\n",
       "      <td>CLC4E_HUMAN; F8WFA1_HUMAN; F5H5X7_HUMAN</td>\n",
       "      <td>ENST00000299663; ENST00000446457, ENST00000450...</td>\n",
       "      <td>ENSP00000299663; ENSP00000387737, ENSP00000404...</td>\n",
       "      <td>ENSG00000166523</td>\n",
       "      <td>...</td>\n",
       "      <td>198</td>\n",
       "      <td>96</td>\n",
       "      <td>90</td>\n",
       "      <td>287</td>\n",
       "      <td>309</td>\n",
       "      <td>94</td>\n",
       "      <td>62</td>\n",
       "      <td>215</td>\n",
       "      <td>261</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000125945</td>\n",
       "      <td>ZNF436</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Zinc finger protein 436</td>\n",
       "      <td>ZNF436, KIAA1710</td>\n",
       "      <td>Q9C0F3</td>\n",
       "      <td>ZN436_HUMAN</td>\n",
       "      <td>ENST00000314011, ENST00000374608, ENST00000635...</td>\n",
       "      <td>ENSP00000313582, ENSP00000363736, ENSP00000489...</td>\n",
       "      <td>ENSG00000125945</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>21</td>\n",
       "      <td>70</td>\n",
       "      <td>39</td>\n",
       "      <td>71</td>\n",
       "      <td>15</td>\n",
       "      <td>26</td>\n",
       "      <td>66</td>\n",
       "      <td>41</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000206145</td>\n",
       "      <td>P2RX6P</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSG00000206145</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3679 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Ensembl gene id     Ensembl gene                     Ensembl biotype  \\\n",
       "0  ENSG00000146555             SDK1                      protein_coding   \n",
       "1  ENSG00000283537  ENSG00000283537  transcribed_unprocessed_pseudogene   \n",
       "2  ENSG00000166523           CLEC4E                      protein_coding   \n",
       "3  ENSG00000125945           ZNF436                      protein_coding   \n",
       "4  ENSG00000206145           P2RX6P  transcribed_unprocessed_pseudogene   \n",
       "\n",
       "                         UniProt proteins  \\\n",
       "0                      Protein sidekick-1   \n",
       "1                                           \n",
       "2  C-type lectin domain family 4 member E   \n",
       "3                 Zinc finger protein 436   \n",
       "4                                           \n",
       "\n",
       "                                    UniProt genes          UniProt accessions  \\\n",
       "0                                            SDK1  Q7Z5N4; F8W6X9; A0A087WTQ6   \n",
       "1                                                                               \n",
       "2  CLEC4E, CLECSF9, MINCLE, UNQ218/PRO244; CLEC4E      Q9ULY5; F8WFA1; F5H5X7   \n",
       "3                                ZNF436, KIAA1710                      Q9C0F3   \n",
       "4                                                                               \n",
       "\n",
       "                                UniProt names  \\\n",
       "0  SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN   \n",
       "1                                               \n",
       "2     CLC4E_HUMAN; F8WFA1_HUMAN; F5H5X7_HUMAN   \n",
       "3                                 ZN436_HUMAN   \n",
       "4                                               \n",
       "\n",
       "                         UniProt Ensembl transcripts  \\\n",
       "0  ENST00000404826; ENST00000389531; ENST00000615806   \n",
       "1                                                      \n",
       "2  ENST00000299663; ENST00000446457, ENST00000450...   \n",
       "3  ENST00000314011, ENST00000374608, ENST00000635...   \n",
       "4                                                      \n",
       "\n",
       "                            UniProt Ensembl proteins UniProt Ensembl gene ids  \\\n",
       "0  ENSP00000385899; ENSP00000374182; ENSP00000478062          ENSG00000146555   \n",
       "1                                                             ENSG00000283537   \n",
       "2  ENSP00000299663; ENSP00000387737, ENSP00000404...          ENSG00000166523   \n",
       "3  ENSP00000313582, ENSP00000363736, ENSP00000489...          ENSG00000125945   \n",
       "4                                                             ENSG00000206145   \n",
       "\n",
       "   ... plt49.run1 pat_3563 plt49.run1 pat_3564 plt49.run1 pat_3565  \\\n",
       "0  ...                   5                   2                   2   \n",
       "1  ...                   0                   0                   1   \n",
       "2  ...                 198                  96                  90   \n",
       "3  ...                  40                  21                  70   \n",
       "4  ...                   0                   0                   0   \n",
       "\n",
       "  plt49.run1 pat_3566 plt49.run1 pat_3567 plt49.run1 pat_3568  \\\n",
       "0                   0                   2                   0   \n",
       "1                   0                   0                   0   \n",
       "2                 287                 309                  94   \n",
       "3                  39                  71                  15   \n",
       "4                   0                   0                   1   \n",
       "\n",
       "  plt49.run1 pat_3569 plt49.run1 pat_3570 plt49.run1 pat_3571  \\\n",
       "0                   1                   2                   1   \n",
       "1                   0                   0                   0   \n",
       "2                  62                 215                 261   \n",
       "3                  26                  66                  41   \n",
       "4                   0                   1                   0   \n",
       "\n",
       "  plt49.run1 pat_3572  \n",
       "0                   0  \n",
       "1                   0  \n",
       "2                 131  \n",
       "3                  17  \n",
       "4                   0  \n",
       "\n",
       "[5 rows x 3679 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.fillna(0)\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df.to_csv('gene_meta.csv', index =False) # Gen-Datei speichern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl gene id</th>\n",
       "      <th>Ensembl gene</th>\n",
       "      <th>Ensembl biotype</th>\n",
       "      <th>UniProt proteins</th>\n",
       "      <th>UniProt genes</th>\n",
       "      <th>UniProt accessions</th>\n",
       "      <th>UniProt names</th>\n",
       "      <th>UniProt Ensembl transcripts</th>\n",
       "      <th>UniProt Ensembl proteins</th>\n",
       "      <th>UniProt Ensembl gene ids</th>\n",
       "      <th>Ensembl chr</th>\n",
       "      <th>Ensembl start</th>\n",
       "      <th>Ensembl stop</th>\n",
       "      <th>Ensembl strand</th>\n",
       "      <th>KEGG PATHWAY terms</th>\n",
       "      <th>KEGG PATHWAY ids</th>\n",
       "      <th>Gene Ontology terms</th>\n",
       "      <th>Gene Ontology ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000146555</td>\n",
       "      <td>SDK1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Protein sidekick-1</td>\n",
       "      <td>SDK1</td>\n",
       "      <td>Q7Z5N4; F8W6X9; A0A087WTQ6</td>\n",
       "      <td>SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN</td>\n",
       "      <td>ENST00000404826; ENST00000389531; ENST00000615806</td>\n",
       "      <td>ENSP00000385899; ENSP00000374182; ENSP00000478062</td>\n",
       "      <td>ENSG00000146555</td>\n",
       "      <td>chr7</td>\n",
       "      <td>3301252</td>\n",
       "      <td>4269000</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>eye development, neural retina development, bi...</td>\n",
       "      <td>GO:0001654, GO:0003407, GO:0005488, GO:0005515...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>chr7</td>\n",
       "      <td>143620943</td>\n",
       "      <td>143645675</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000166523</td>\n",
       "      <td>CLEC4E</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>C-type lectin domain family 4 member E</td>\n",
       "      <td>CLEC4E, CLECSF9, MINCLE, UNQ218/PRO244; CLEC4E</td>\n",
       "      <td>Q9ULY5; F8WFA1; F5H5X7</td>\n",
       "      <td>CLC4E_HUMAN; F8WFA1_HUMAN; F5H5X7_HUMAN</td>\n",
       "      <td>ENST00000299663; ENST00000446457, ENST00000450...</td>\n",
       "      <td>ENSP00000299663; ENSP00000387737, ENSP00000404...</td>\n",
       "      <td>ENSG00000166523</td>\n",
       "      <td>chr12</td>\n",
       "      <td>8533305</td>\n",
       "      <td>8540905</td>\n",
       "      <td>-</td>\n",
       "      <td>Tuberculosis</td>\n",
       "      <td>hsa05152</td>\n",
       "      <td>cell activation, cytokine production, regulati...</td>\n",
       "      <td>GO:0001775, GO:0001816, GO:0001817, GO:0001819...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000125945</td>\n",
       "      <td>ZNF436</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Zinc finger protein 436</td>\n",
       "      <td>ZNF436, KIAA1710</td>\n",
       "      <td>Q9C0F3</td>\n",
       "      <td>ZN436_HUMAN</td>\n",
       "      <td>ENST00000314011, ENST00000374608, ENST00000635...</td>\n",
       "      <td>ENSP00000313582, ENSP00000363736, ENSP00000489...</td>\n",
       "      <td>ENSG00000125945</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23359448</td>\n",
       "      <td>23369836</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>nucleic acid binding transcription factor acti...</td>\n",
       "      <td>GO:0001071, GO:0003676, GO:0003677, GO:0003700...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000206145</td>\n",
       "      <td>P2RX6P</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSG00000206145</td>\n",
       "      <td>chr22</td>\n",
       "      <td>21035243</td>\n",
       "      <td>21045017</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Ensembl gene id     Ensembl gene                     Ensembl biotype  \\\n",
       "0  ENSG00000146555             SDK1                      protein_coding   \n",
       "1  ENSG00000283537  ENSG00000283537  transcribed_unprocessed_pseudogene   \n",
       "2  ENSG00000166523           CLEC4E                      protein_coding   \n",
       "3  ENSG00000125945           ZNF436                      protein_coding   \n",
       "4  ENSG00000206145           P2RX6P  transcribed_unprocessed_pseudogene   \n",
       "\n",
       "                         UniProt proteins  \\\n",
       "0                      Protein sidekick-1   \n",
       "1                                           \n",
       "2  C-type lectin domain family 4 member E   \n",
       "3                 Zinc finger protein 436   \n",
       "4                                           \n",
       "\n",
       "                                    UniProt genes          UniProt accessions  \\\n",
       "0                                            SDK1  Q7Z5N4; F8W6X9; A0A087WTQ6   \n",
       "1                                                                               \n",
       "2  CLEC4E, CLECSF9, MINCLE, UNQ218/PRO244; CLEC4E      Q9ULY5; F8WFA1; F5H5X7   \n",
       "3                                ZNF436, KIAA1710                      Q9C0F3   \n",
       "4                                                                               \n",
       "\n",
       "                                UniProt names  \\\n",
       "0  SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN   \n",
       "1                                               \n",
       "2     CLC4E_HUMAN; F8WFA1_HUMAN; F5H5X7_HUMAN   \n",
       "3                                 ZN436_HUMAN   \n",
       "4                                               \n",
       "\n",
       "                         UniProt Ensembl transcripts  \\\n",
       "0  ENST00000404826; ENST00000389531; ENST00000615806   \n",
       "1                                                      \n",
       "2  ENST00000299663; ENST00000446457, ENST00000450...   \n",
       "3  ENST00000314011, ENST00000374608, ENST00000635...   \n",
       "4                                                      \n",
       "\n",
       "                            UniProt Ensembl proteins UniProt Ensembl gene ids  \\\n",
       "0  ENSP00000385899; ENSP00000374182; ENSP00000478062          ENSG00000146555   \n",
       "1                                                             ENSG00000283537   \n",
       "2  ENSP00000299663; ENSP00000387737, ENSP00000404...          ENSG00000166523   \n",
       "3  ENSP00000313582, ENSP00000363736, ENSP00000489...          ENSG00000125945   \n",
       "4                                                             ENSG00000206145   \n",
       "\n",
       "  Ensembl chr Ensembl start Ensembl stop Ensembl strand KEGG PATHWAY terms  \\\n",
       "0        chr7       3301252      4269000              +                      \n",
       "1        chr7     143620943    143645675              +                      \n",
       "2       chr12       8533305      8540905              -       Tuberculosis   \n",
       "3        chr1      23359448     23369836              -                      \n",
       "4       chr22      21035243     21045017              -                      \n",
       "\n",
       "  KEGG PATHWAY ids                                Gene Ontology terms  \\\n",
       "0                   eye development, neural retina development, bi...   \n",
       "1                                                                       \n",
       "2         hsa05152  cell activation, cytokine production, regulati...   \n",
       "3                   nucleic acid binding transcription factor acti...   \n",
       "4                                                                       \n",
       "\n",
       "                                   Gene Ontology ids  \n",
       "0  GO:0001654, GO:0003407, GO:0005488, GO:0005515...  \n",
       "1                                                     \n",
       "2  GO:0001775, GO:0001816, GO:0001817, GO:0001819...  \n",
       "3  GO:0001071, GO:0003676, GO:0003677, GO:0003700...  \n",
       "4                                                     "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features = gene_df.loc[:,['Ensembl gene id', 'Ensembl gene', 'Ensembl biotype', 'UniProt proteins', 'UniProt genes', 'UniProt accessions', 'UniProt names', 'UniProt Ensembl transcripts', 'UniProt Ensembl proteins', 'UniProt Ensembl gene ids', 'Ensembl chr', 'Ensembl start', 'Ensembl stop', 'Ensembl strand', 'KEGG PATHWAY terms', 'KEGG PATHWAY ids', 'Gene Ontology terms', 'Gene Ontology ids']] # Auswahl der ersten 18 Spalten -> die Gene ids und Spalten mit Metadaten\n",
    "features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "features.to_csv(\"variables.tsv\", sep=\"\\t\", index=False, header=False) # Variables-Datei speichern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. --- Barcodes Datei aus Patienten-Metadaten erstellen ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['plt1.run1 pat_88', 'plt1.run1 pat_89', 'plt1.run1 pat_90',\n",
       "       'plt1.run1 pat_91', 'plt1.run1 pat_92', 'plt1.run1 pat_93',\n",
       "       'plt1.run1 pat_94', 'plt1.run1 pat_95', 'plt1.run1 pat_96',\n",
       "       'plt1.run1 pat_97',\n",
       "       ...\n",
       "       'plt49.run1 pat_3563', 'plt49.run1 pat_3564', 'plt49.run1 pat_3565',\n",
       "       'plt49.run1 pat_3566', 'plt49.run1 pat_3567', 'plt49.run1 pat_3568',\n",
       "       'plt49.run1 pat_3569', 'plt49.run1 pat_3570', 'plt49.run1 pat_3571',\n",
       "       'plt49.run1 pat_3572'],\n",
       "      dtype='object', length=3661)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Barcodes-Datei (Patienten + Metadaten)\n",
    "\n",
    "patient_ids = gene_df.columns[18:]  # Patienten-Spalten starten ab Spalte 18 \n",
    "patient_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rna project.run id</th>\n",
       "      <th>TubeID</th>\n",
       "      <th>meta TubeID</th>\n",
       "      <th>meta pop export_psn</th>\n",
       "      <th>meta pop visit_1_date</th>\n",
       "      <th>meta pop visit_2_date</th>\n",
       "      <th>meta pop visit_3_date</th>\n",
       "      <th>meta pop visit_4_date</th>\n",
       "      <th>meta pop age</th>\n",
       "      <th>meta pop age_dec</th>\n",
       "      <th>...</th>\n",
       "      <th>rna unassigned reads not mapping</th>\n",
       "      <th>rna unassigned reads due to bad mapping quality</th>\n",
       "      <th>rna unassigned reads due to low fragment length</th>\n",
       "      <th>rna unassigned chimeric reads</th>\n",
       "      <th>rna usable sample: mapped reads &gt; 2000000</th>\n",
       "      <th>rna DDX3Y</th>\n",
       "      <th>rna XIST</th>\n",
       "      <th>rna gender</th>\n",
       "      <th>rna sample id original</th>\n",
       "      <th>rna plate id original</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3656</th>\n",
       "      <td>plt49.run1 pat_3568</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3657</th>\n",
       "      <td>plt49.run1 pat_3569</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3658</th>\n",
       "      <td>plt49.run1 pat_3570</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3659</th>\n",
       "      <td>plt49.run1 pat_3571</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3660</th>\n",
       "      <td>plt49.run1 pat_3572</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       rna project.run id TubeID meta TubeID meta pop export_psn  \\\n",
       "3656  plt49.run1 pat_3568      0           0                   0   \n",
       "3657  plt49.run1 pat_3569      0           0                   0   \n",
       "3658  plt49.run1 pat_3570      0           0                   0   \n",
       "3659  plt49.run1 pat_3571      0           0                   0   \n",
       "3660  plt49.run1 pat_3572      0           0                   0   \n",
       "\n",
       "     meta pop visit_1_date meta pop visit_2_date meta pop visit_3_date  \\\n",
       "3656                     0                     0                     0   \n",
       "3657                     0                     0                     0   \n",
       "3658                     0                     0                     0   \n",
       "3659                     0                     0                     0   \n",
       "3660                     0                     0                     0   \n",
       "\n",
       "     meta pop visit_4_date  meta pop age meta pop age_dec  ...  \\\n",
       "3656                     0           0.0                0  ...   \n",
       "3657                     0           0.0                0  ...   \n",
       "3658                     0           0.0                0  ...   \n",
       "3659                     0           0.0                0  ...   \n",
       "3660                     0           0.0                0  ...   \n",
       "\n",
       "     rna unassigned reads not mapping  \\\n",
       "3656                              0.0   \n",
       "3657                              0.0   \n",
       "3658                              0.0   \n",
       "3659                              0.0   \n",
       "3660                              0.0   \n",
       "\n",
       "     rna unassigned reads due to bad mapping quality  \\\n",
       "3656                                             0.0   \n",
       "3657                                             0.0   \n",
       "3658                                             0.0   \n",
       "3659                                             0.0   \n",
       "3660                                             0.0   \n",
       "\n",
       "     rna unassigned reads due to low fragment length  \\\n",
       "3656                                             0.0   \n",
       "3657                                             0.0   \n",
       "3658                                             0.0   \n",
       "3659                                             0.0   \n",
       "3660                                             0.0   \n",
       "\n",
       "     rna unassigned chimeric reads rna usable sample: mapped reads > 2000000  \\\n",
       "3656                           0.0                                         0   \n",
       "3657                           0.0                                         0   \n",
       "3658                           0.0                                         0   \n",
       "3659                           0.0                                         0   \n",
       "3660                           0.0                                         0   \n",
       "\n",
       "     rna DDX3Y rna XIST rna gender rna sample id original  \\\n",
       "3656       0.0      0.0          0                      0   \n",
       "3657       0.0      0.0          0                      0   \n",
       "3658       0.0      0.0          0                      0   \n",
       "3659       0.0      0.0          0                      0   \n",
       "3660       0.0      0.0          0                      0   \n",
       "\n",
       "     rna plate id original  \n",
       "3656                     0  \n",
       "3657                     0  \n",
       "3658                     0  \n",
       "3659                     0  \n",
       "3660                     0  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes = pd.DataFrame({\n",
    "    \"rna project.run id\": patient_ids # Benennung der Spalte in barcodes.tsv mit \"rna project.run id\"\n",
    "}).merge(pat_df, left_on=\"rna project.run id\", right_on=\"rna project.run id\", how=\"left\")  # Metadaten anfügen\n",
    "barcodes = barcodes.fillna(0)\n",
    "barcodes.to_csv(\"barcodes.tsv\", sep=\"\\t\", index=False, header=False)\n",
    "barcodes.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes = pd.read_csv(\"barcodes.tsv\", decimal=\".\", sep=\"\\t\") # Barcodes-Datei speichern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. --- Count-Matrix erstellen aus Gen-Metadaten ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>plt1.run1 pat_88</th>\n",
       "      <th>plt1.run1 pat_89</th>\n",
       "      <th>plt1.run1 pat_90</th>\n",
       "      <th>plt1.run1 pat_91</th>\n",
       "      <th>plt1.run1 pat_92</th>\n",
       "      <th>plt1.run1 pat_93</th>\n",
       "      <th>plt1.run1 pat_94</th>\n",
       "      <th>plt1.run1 pat_95</th>\n",
       "      <th>plt1.run1 pat_96</th>\n",
       "      <th>plt1.run1 pat_97</th>\n",
       "      <th>...</th>\n",
       "      <th>plt49.run1 pat_3563</th>\n",
       "      <th>plt49.run1 pat_3564</th>\n",
       "      <th>plt49.run1 pat_3565</th>\n",
       "      <th>plt49.run1 pat_3566</th>\n",
       "      <th>plt49.run1 pat_3567</th>\n",
       "      <th>plt49.run1 pat_3568</th>\n",
       "      <th>plt49.run1 pat_3569</th>\n",
       "      <th>plt49.run1 pat_3570</th>\n",
       "      <th>plt49.run1 pat_3571</th>\n",
       "      <th>plt49.run1 pat_3572</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>225</td>\n",
       "      <td>34</td>\n",
       "      <td>13</td>\n",
       "      <td>102</td>\n",
       "      <td>156</td>\n",
       "      <td>175</td>\n",
       "      <td>389</td>\n",
       "      <td>450</td>\n",
       "      <td>224</td>\n",
       "      <td>157</td>\n",
       "      <td>...</td>\n",
       "      <td>198</td>\n",
       "      <td>96</td>\n",
       "      <td>90</td>\n",
       "      <td>287</td>\n",
       "      <td>309</td>\n",
       "      <td>94</td>\n",
       "      <td>62</td>\n",
       "      <td>215</td>\n",
       "      <td>261</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>73</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>48</td>\n",
       "      <td>102</td>\n",
       "      <td>27</td>\n",
       "      <td>65</td>\n",
       "      <td>62</td>\n",
       "      <td>90</td>\n",
       "      <td>73</td>\n",
       "      <td>...</td>\n",
       "      <td>40</td>\n",
       "      <td>21</td>\n",
       "      <td>70</td>\n",
       "      <td>39</td>\n",
       "      <td>71</td>\n",
       "      <td>15</td>\n",
       "      <td>26</td>\n",
       "      <td>66</td>\n",
       "      <td>41</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57026</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57027</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57028</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57029</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57030</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57031 rows × 3661 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      plt1.run1 pat_88 plt1.run1 pat_89 plt1.run1 pat_90 plt1.run1 pat_91  \\\n",
       "0                    2                0                0                3   \n",
       "1                    0                0                0                0   \n",
       "2                  225               34               13              102   \n",
       "3                   73               17                2               48   \n",
       "4                    0                0                0                0   \n",
       "...                ...              ...              ...              ...   \n",
       "57026                0                0                0                0   \n",
       "57027                0                0                0                0   \n",
       "57028                0                0                0                0   \n",
       "57029                0                0                0                0   \n",
       "57030                0                0                0                0   \n",
       "\n",
       "      plt1.run1 pat_92 plt1.run1 pat_93 plt1.run1 pat_94 plt1.run1 pat_95  \\\n",
       "0                    3                0                1                1   \n",
       "1                    0                0                0                0   \n",
       "2                  156              175              389              450   \n",
       "3                  102               27               65               62   \n",
       "4                    4                0                2                1   \n",
       "...                ...              ...              ...              ...   \n",
       "57026                0                0                0                0   \n",
       "57027                0                0                0                0   \n",
       "57028                0                0                0                0   \n",
       "57029                0                0                0                0   \n",
       "57030                0                0                0                0   \n",
       "\n",
       "      plt1.run1 pat_96 plt1.run1 pat_97  ... plt49.run1 pat_3563  \\\n",
       "0                   16               11  ...                   5   \n",
       "1                    0                3  ...                   0   \n",
       "2                  224              157  ...                 198   \n",
       "3                   90               73  ...                  40   \n",
       "4                    0                0  ...                   0   \n",
       "...                ...              ...  ...                 ...   \n",
       "57026                0                0  ...                   0   \n",
       "57027                0                0  ...                   0   \n",
       "57028                0                0  ...                   0   \n",
       "57029                0                0  ...                   0   \n",
       "57030                0                0  ...                   0   \n",
       "\n",
       "      plt49.run1 pat_3564 plt49.run1 pat_3565 plt49.run1 pat_3566  \\\n",
       "0                       2                   2                   0   \n",
       "1                       0                   1                   0   \n",
       "2                      96                  90                 287   \n",
       "3                      21                  70                  39   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "57026                   0                   0                   0   \n",
       "57027                   0                   0                   0   \n",
       "57028                   0                   0                   0   \n",
       "57029                   0                   0                   0   \n",
       "57030                   0                   0                   0   \n",
       "\n",
       "      plt49.run1 pat_3567 plt49.run1 pat_3568 plt49.run1 pat_3569  \\\n",
       "0                       2                   0                   1   \n",
       "1                       0                   0                   0   \n",
       "2                     309                  94                  62   \n",
       "3                      71                  15                  26   \n",
       "4                       0                   1                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "57026                   0                   0                   0   \n",
       "57027                   0                   0                   0   \n",
       "57028                   0                   0                   0   \n",
       "57029                   0                   0                   0   \n",
       "57030                   0                   0                   0   \n",
       "\n",
       "      plt49.run1 pat_3570 plt49.run1 pat_3571 plt49.run1 pat_3572  \n",
       "0                       2                   1                   0  \n",
       "1                       0                   0                   0  \n",
       "2                     215                 261                 131  \n",
       "3                      66                  41                  17  \n",
       "4                       1                   0                   0  \n",
       "...                   ...                 ...                 ...  \n",
       "57026                   0                   0                   0  \n",
       "57027                   0                   0                   0  \n",
       "57028                   0                   0                   0  \n",
       "57029                   0                   0                   0  \n",
       "57030                   0                   0                   0  \n",
       "\n",
       "[57031 rows x 3661 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = gene_df.iloc[:, 18:].fillna(0) # Zählwerte extrahieren und NaN durch 0 ersetzen\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<COOrdinate sparse matrix of dtype 'object'\n",
       "\twith 208790491 stored elements and shape (57031, 3661)>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sparse_matrix = coo_matrix(counts.values) # Sparse-Matrix erstellen\n",
    "sparse_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "mmwrite(\"matrix.mtx\", sparse_matrix) # MTX-Datei speichern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. --- Anndata-Objekt aus .mtx, barcodes.tsv und genes.tsv erstellen ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx_tsv = 'matrix.mtx' # Datei, die die Counts enthält\n",
    "barcodes_tsv = 'barcodes.tsv' # Datei, die Barcode-Informationen enthält\n",
    "variables_tsv = 'variables.tsv' # Datei, die variable Informationen enthält"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3776, 75) (3660, 75) (57031, 3679) (57031, 18)\n"
     ]
    }
   ],
   "source": [
    "print(pat_df.shape, barcodes.shape, gene_df.shape, features.shape) # Shape überprüfen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] Reading files: 1 of 1 \n"
     ]
    }
   ],
   "source": [
    "if path_mtx:\n",
    "    adata = assembler.from_mtx(path_mtx, mtx=mtx_tsv, barcodes=barcodes_tsv, variables=variables_tsv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6. --- AnnData vorbereiten ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3661 × 57031\n",
       "    obs: 'TubeID', 'meta TubeID', 'meta pop export_psn', 'meta pop visit_1_date', 'meta pop visit_2_date', 'meta pop visit_3_date', 'meta pop visit_4_date', 'meta pop age', 'meta pop age_dec', 'meta pop sex', 'meta pop c19_severity', 'meta pop cohort', 'meta suep_hap export_psn', 'meta suep_hap baseline_date', 'meta suep_hap end_acute_date', 'meta suep_hap end_acute_visit_date', 'meta suep_hap m3_fu_date', 'meta suep_hap m6_fu_date', 'meta suep_hap m12_fu_date', 'meta suep_hap m24_fu_date', 'meta suep_hap age', 'meta suep_hap age_dec', 'meta suep_hap sex', 'meta suep_hap c19_severity', 'meta suep_hap cohort', 'meta export_psn', 'meta age', 'meta age_dec', 'meta sex', 'meta c19_severity', 'meta cohort', 'rna TubeID', 'rna id', 'rna fastq1', 'rna project', 'rna run', 'rna plate', 'rna organism', 'rna assembly', 'rna release', 'rna parameters', 'rna reads processed', 'rna reads with exactly 1 alignment', 'rna reads with multiple alignments', 'rna reads with no alignment', 'rna reads aligned', 'rna % reads aligned', 'rna duplicate reads', 'rna % duplicate reads', 'rna mitochondrial reads', 'rna % mitochondrial reads', 'rna rrna subunit reads', 'rna % rrna subunit reads', 'rna pcr bottleneck coefficient', 'rna mean fragment size (only for paired-end)', 'rna reads used for further steps (after optional filters for multimap, duplicate, mitochondria, rrna)', 'rna parameters.1', 'rna reads processed.1', 'rna reads assigned', 'rna % reads assigned total', 'rna % features with cpm >= 0.5', 'rna unassigned reads overlapping multiple genes', 'rna unassigned reads mapping to multiple locations', 'rna unassigned reads mapping to no feature', 'rna unassigned reads not mapping', 'rna unassigned reads due to bad mapping quality', 'rna unassigned reads due to low fragment length', 'rna unassigned chimeric reads', 'rna usable sample: mapped reads > 2000000', 'rna DDX3Y', 'rna XIST', 'rna gender', 'rna sample id original', 'rna plate id original', 'filename', 'rel_path'\n",
       "    var: 'Ensembl gene', 'Ensembl biotype', 'UniProt proteins', 'UniProt genes', 'UniProt accessions', 'UniProt names', 'UniProt Ensembl transcripts', 'UniProt Ensembl proteins', 'UniProt Ensembl gene ids', 'Ensembl chr', 'Ensembl start', 'Ensembl stop', 'Ensembl strand', 'KEGG PATHWAY terms', 'KEGG PATHWAY ids', 'Gene Ontology terms', 'Gene Ontology ids'"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Adata Obs einen Header hinzufügren\n",
    "adata.obs_names.name = 'rna project.run id'\n",
    "obs_columns_list = pat_df.columns.tolist()\n",
    "obs_columns_list.remove('rna project.run id')\n",
    "obs_columns_list.append('filename') # möglicherweise muss die komplette Spalte raus\n",
    "obs_columns_list.append('rel_path') # möglicherweise muss die komplette Spalte raus\n",
    "obs_header = obs_columns_list\n",
    "adata.obs.columns = obs_header\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_header = features.columns.tolist()\n",
    "var_header.remove(\"Ensembl gene id\")\n",
    "adata.var_names.name = \"Ensembl gene id\"\n",
    "adata.var.columns = var_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TubeID</th>\n",
       "      <th>meta TubeID</th>\n",
       "      <th>meta pop export_psn</th>\n",
       "      <th>meta pop visit_1_date</th>\n",
       "      <th>meta pop visit_2_date</th>\n",
       "      <th>meta pop visit_3_date</th>\n",
       "      <th>meta pop visit_4_date</th>\n",
       "      <th>meta pop age</th>\n",
       "      <th>meta pop age_dec</th>\n",
       "      <th>meta pop sex</th>\n",
       "      <th>meta pop c19_severity</th>\n",
       "      <th>meta pop cohort</th>\n",
       "      <th>meta suep_hap export_psn</th>\n",
       "      <th>meta suep_hap baseline_date</th>\n",
       "      <th>meta suep_hap end_acute_date</th>\n",
       "      <th>meta suep_hap end_acute_visit_date</th>\n",
       "      <th>meta suep_hap m3_fu_date</th>\n",
       "      <th>meta suep_hap m6_fu_date</th>\n",
       "      <th>meta suep_hap m12_fu_date</th>\n",
       "      <th>meta suep_hap m24_fu_date</th>\n",
       "      <th>meta suep_hap age</th>\n",
       "      <th>meta suep_hap age_dec</th>\n",
       "      <th>meta suep_hap sex</th>\n",
       "      <th>meta suep_hap c19_severity</th>\n",
       "      <th>meta suep_hap cohort</th>\n",
       "      <th>meta export_psn</th>\n",
       "      <th>meta age</th>\n",
       "      <th>meta age_dec</th>\n",
       "      <th>meta sex</th>\n",
       "      <th>meta c19_severity</th>\n",
       "      <th>meta cohort</th>\n",
       "      <th>rna TubeID</th>\n",
       "      <th>rna id</th>\n",
       "      <th>rna fastq1</th>\n",
       "      <th>rna project</th>\n",
       "      <th>rna run</th>\n",
       "      <th>rna plate</th>\n",
       "      <th>rna organism</th>\n",
       "      <th>rna assembly</th>\n",
       "      <th>rna release</th>\n",
       "      <th>rna parameters</th>\n",
       "      <th>rna reads processed</th>\n",
       "      <th>rna reads with exactly 1 alignment</th>\n",
       "      <th>rna reads with multiple alignments</th>\n",
       "      <th>rna reads with no alignment</th>\n",
       "      <th>rna reads aligned</th>\n",
       "      <th>rna % reads aligned</th>\n",
       "      <th>rna duplicate reads</th>\n",
       "      <th>rna % duplicate reads</th>\n",
       "      <th>rna mitochondrial reads</th>\n",
       "      <th>rna % mitochondrial reads</th>\n",
       "      <th>rna rrna subunit reads</th>\n",
       "      <th>rna % rrna subunit reads</th>\n",
       "      <th>rna pcr bottleneck coefficient</th>\n",
       "      <th>rna mean fragment size (only for paired-end)</th>\n",
       "      <th>rna reads used for further steps (after optional filters for multimap, duplicate, mitochondria, rrna)</th>\n",
       "      <th>rna parameters.1</th>\n",
       "      <th>rna reads processed.1</th>\n",
       "      <th>rna reads assigned</th>\n",
       "      <th>rna % reads assigned total</th>\n",
       "      <th>rna % features with cpm &gt;= 0.5</th>\n",
       "      <th>rna unassigned reads overlapping multiple genes</th>\n",
       "      <th>rna unassigned reads mapping to multiple locations</th>\n",
       "      <th>rna unassigned reads mapping to no feature</th>\n",
       "      <th>rna unassigned reads not mapping</th>\n",
       "      <th>rna unassigned reads due to bad mapping quality</th>\n",
       "      <th>rna unassigned reads due to low fragment length</th>\n",
       "      <th>rna unassigned chimeric reads</th>\n",
       "      <th>rna usable sample: mapped reads &gt; 2000000</th>\n",
       "      <th>rna DDX3Y</th>\n",
       "      <th>rna XIST</th>\n",
       "      <th>rna gender</th>\n",
       "      <th>rna sample id original</th>\n",
       "      <th>rna plate id original</th>\n",
       "      <th>filename</th>\n",
       "      <th>rel_path</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rna project.run id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>plt1.run1 pat_88</th>\n",
       "      <td>LV1007699778</td>\n",
       "      <td>LV1007699778</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>hap_0178</td>\n",
       "      <td>17/11/2021</td>\n",
       "      <td>24/11/2021</td>\n",
       "      <td>24/11/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>60-69</td>\n",
       "      <td>M</td>\n",
       "      <td>severe</td>\n",
       "      <td>hap</td>\n",
       "      <td>hap_0178</td>\n",
       "      <td>65.0</td>\n",
       "      <td>60-69</td>\n",
       "      <td>M</td>\n",
       "      <td>severe</td>\n",
       "      <td>hap</td>\n",
       "      <td>LV1007699778</td>\n",
       "      <td>pat_88</td>\n",
       "      <td>AAAVY52HV_NAPKON_Plate1_2_22s004137-1-1_Valasa...</td>\n",
       "      <td>plt1</td>\n",
       "      <td>run1</td>\n",
       "      <td>plate1</td>\n",
       "      <td>human</td>\n",
       "      <td>hg38</td>\n",
       "      <td>104.0</td>\n",
       "      <td>mode: single-end, keep duplicates: no, keep mu...</td>\n",
       "      <td>29,112,743</td>\n",
       "      <td>23,152,813</td>\n",
       "      <td>4,044,454</td>\n",
       "      <td>1,915,476</td>\n",
       "      <td>27,197,267</td>\n",
       "      <td>93.0</td>\n",
       "      <td>20,093,139</td>\n",
       "      <td>73.0</td>\n",
       "      <td>1,714,158</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1,868,907</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6,644,895</td>\n",
       "      <td>-t exon -g gene_id -s 2; multi-mapping: no; du...</td>\n",
       "      <td>6,644,895</td>\n",
       "      <td>4,810,861</td>\n",
       "      <td>16.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>216.360</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1,617,674</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>844.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Sample 49</td>\n",
       "      <td>Plate 1</td>\n",
       "      <td>matrix.mtx</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plt1.run1 pat_89</th>\n",
       "      <td>LV1007699743</td>\n",
       "      <td>LV1007699743</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>suep_0425</td>\n",
       "      <td>08/11/2021</td>\n",
       "      <td>17/11/2021</td>\n",
       "      <td>17/11/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>60-69</td>\n",
       "      <td>M</td>\n",
       "      <td>mild</td>\n",
       "      <td>suep</td>\n",
       "      <td>suep_0425</td>\n",
       "      <td>61.0</td>\n",
       "      <td>60-69</td>\n",
       "      <td>M</td>\n",
       "      <td>mild</td>\n",
       "      <td>suep</td>\n",
       "      <td>LV1007699743</td>\n",
       "      <td>pat_89</td>\n",
       "      <td>AAAVY52HV_NAPKON_Plate1_2_22s004137-1-1_Valasa...</td>\n",
       "      <td>plt1</td>\n",
       "      <td>run1</td>\n",
       "      <td>plate1</td>\n",
       "      <td>human</td>\n",
       "      <td>hg38</td>\n",
       "      <td>104.0</td>\n",
       "      <td>mode: single-end, keep duplicates: no, keep mu...</td>\n",
       "      <td>24,355,220</td>\n",
       "      <td>13,253,462</td>\n",
       "      <td>7,713,379</td>\n",
       "      <td>3,388,379</td>\n",
       "      <td>20,966,841</td>\n",
       "      <td>86.0</td>\n",
       "      <td>17,678,467</td>\n",
       "      <td>84.0</td>\n",
       "      <td>1,475,002</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6,372,269</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2,989,995</td>\n",
       "      <td>-t exon -g gene_id -s 2; multi-mapping: no; du...</td>\n",
       "      <td>2,989,995</td>\n",
       "      <td>2,155,768</td>\n",
       "      <td>8.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>97.961</td>\n",
       "      <td>0.0</td>\n",
       "      <td>736,266</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>670.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Sample 50</td>\n",
       "      <td>Plate 1</td>\n",
       "      <td>matrix.mtx</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plt49.run1 pat_3571</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>matrix.mtx</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plt49.run1 pat_3572</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>matrix.mtx</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3661 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           TubeID   meta TubeID meta pop export_psn  \\\n",
       "rna project.run id                                                    \n",
       "plt1.run1 pat_88     LV1007699778  LV1007699778                   0   \n",
       "plt1.run1 pat_89     LV1007699743  LV1007699743                   0   \n",
       "...                           ...           ...                 ...   \n",
       "plt49.run1 pat_3571             0             0                   0   \n",
       "plt49.run1 pat_3572             0             0                   0   \n",
       "\n",
       "                    meta pop visit_1_date meta pop visit_2_date  \\\n",
       "rna project.run id                                                \n",
       "plt1.run1 pat_88                        0                     0   \n",
       "plt1.run1 pat_89                        0                     0   \n",
       "...                                   ...                   ...   \n",
       "plt49.run1 pat_3571                     0                     0   \n",
       "plt49.run1 pat_3572                     0                     0   \n",
       "\n",
       "                    meta pop visit_3_date meta pop visit_4_date  meta pop age  \\\n",
       "rna project.run id                                                              \n",
       "plt1.run1 pat_88                        0                     0           0.0   \n",
       "plt1.run1 pat_89                        0                     0           0.0   \n",
       "...                                   ...                   ...           ...   \n",
       "plt49.run1 pat_3571                     0                     0           0.0   \n",
       "plt49.run1 pat_3572                     0                     0           0.0   \n",
       "\n",
       "                    meta pop age_dec meta pop sex meta pop c19_severity  \\\n",
       "rna project.run id                                                        \n",
       "plt1.run1 pat_88                   0            0                     0   \n",
       "plt1.run1 pat_89                   0            0                     0   \n",
       "...                              ...          ...                   ...   \n",
       "plt49.run1 pat_3571                0            0                     0   \n",
       "plt49.run1 pat_3572                0            0                     0   \n",
       "\n",
       "                    meta pop cohort meta suep_hap export_psn  \\\n",
       "rna project.run id                                             \n",
       "plt1.run1 pat_88                  0                 hap_0178   \n",
       "plt1.run1 pat_89                  0                suep_0425   \n",
       "...                             ...                      ...   \n",
       "plt49.run1 pat_3571               0                        0   \n",
       "plt49.run1 pat_3572               0                        0   \n",
       "\n",
       "                    meta suep_hap baseline_date meta suep_hap end_acute_date  \\\n",
       "rna project.run id                                                             \n",
       "plt1.run1 pat_88                     17/11/2021                   24/11/2021   \n",
       "plt1.run1 pat_89                     08/11/2021                   17/11/2021   \n",
       "...                                         ...                          ...   \n",
       "plt49.run1 pat_3571                           0                            0   \n",
       "plt49.run1 pat_3572                           0                            0   \n",
       "\n",
       "                    meta suep_hap end_acute_visit_date  \\\n",
       "rna project.run id                                       \n",
       "plt1.run1 pat_88                            24/11/2021   \n",
       "plt1.run1 pat_89                            17/11/2021   \n",
       "...                                                ...   \n",
       "plt49.run1 pat_3571                                  0   \n",
       "plt49.run1 pat_3572                                  0   \n",
       "\n",
       "                    meta suep_hap m3_fu_date meta suep_hap m6_fu_date  \\\n",
       "rna project.run id                                                      \n",
       "plt1.run1 pat_88                           0                        0   \n",
       "plt1.run1 pat_89                           0                        0   \n",
       "...                                      ...                      ...   \n",
       "plt49.run1 pat_3571                        0                        0   \n",
       "plt49.run1 pat_3572                        0                        0   \n",
       "\n",
       "                    meta suep_hap m12_fu_date meta suep_hap m24_fu_date  \\\n",
       "rna project.run id                                                        \n",
       "plt1.run1 pat_88                            0                         0   \n",
       "plt1.run1 pat_89                            0                         0   \n",
       "...                                       ...                       ...   \n",
       "plt49.run1 pat_3571                         0                         0   \n",
       "plt49.run1 pat_3572                         0                         0   \n",
       "\n",
       "                     meta suep_hap age meta suep_hap age_dec  \\\n",
       "rna project.run id                                             \n",
       "plt1.run1 pat_88                  65.0                 60-69   \n",
       "plt1.run1 pat_89                  61.0                 60-69   \n",
       "...                                ...                   ...   \n",
       "plt49.run1 pat_3571                0.0                     0   \n",
       "plt49.run1 pat_3572                0.0                     0   \n",
       "\n",
       "                    meta suep_hap sex meta suep_hap c19_severity  \\\n",
       "rna project.run id                                                 \n",
       "plt1.run1 pat_88                    M                     severe   \n",
       "plt1.run1 pat_89                    M                       mild   \n",
       "...                               ...                        ...   \n",
       "plt49.run1 pat_3571                 0                          0   \n",
       "plt49.run1 pat_3572                 0                          0   \n",
       "\n",
       "                    meta suep_hap cohort meta export_psn  meta age  \\\n",
       "rna project.run id                                                   \n",
       "plt1.run1 pat_88                     hap        hap_0178      65.0   \n",
       "plt1.run1 pat_89                    suep       suep_0425      61.0   \n",
       "...                                  ...             ...       ...   \n",
       "plt49.run1 pat_3571                    0               0       0.0   \n",
       "plt49.run1 pat_3572                    0               0       0.0   \n",
       "\n",
       "                    meta age_dec meta sex meta c19_severity meta cohort  \\\n",
       "rna project.run id                                                        \n",
       "plt1.run1 pat_88           60-69        M            severe         hap   \n",
       "plt1.run1 pat_89           60-69        M              mild        suep   \n",
       "...                          ...      ...               ...         ...   \n",
       "plt49.run1 pat_3571            0        0                 0           0   \n",
       "plt49.run1 pat_3572            0        0                 0           0   \n",
       "\n",
       "                       rna TubeID  rna id  \\\n",
       "rna project.run id                          \n",
       "plt1.run1 pat_88     LV1007699778  pat_88   \n",
       "plt1.run1 pat_89     LV1007699743  pat_89   \n",
       "...                           ...     ...   \n",
       "plt49.run1 pat_3571             0       0   \n",
       "plt49.run1 pat_3572             0       0   \n",
       "\n",
       "                                                            rna fastq1  \\\n",
       "rna project.run id                                                       \n",
       "plt1.run1 pat_88     AAAVY52HV_NAPKON_Plate1_2_22s004137-1-1_Valasa...   \n",
       "plt1.run1 pat_89     AAAVY52HV_NAPKON_Plate1_2_22s004137-1-1_Valasa...   \n",
       "...                                                                ...   \n",
       "plt49.run1 pat_3571                                                  0   \n",
       "plt49.run1 pat_3572                                                  0   \n",
       "\n",
       "                    rna project rna run rna plate rna organism rna assembly  \\\n",
       "rna project.run id                                                            \n",
       "plt1.run1 pat_88           plt1    run1    plate1        human         hg38   \n",
       "plt1.run1 pat_89           plt1    run1    plate1        human         hg38   \n",
       "...                         ...     ...       ...          ...          ...   \n",
       "plt49.run1 pat_3571           0       0         0            0            0   \n",
       "plt49.run1 pat_3572           0       0         0            0            0   \n",
       "\n",
       "                     rna release  \\\n",
       "rna project.run id                 \n",
       "plt1.run1 pat_88           104.0   \n",
       "plt1.run1 pat_89           104.0   \n",
       "...                          ...   \n",
       "plt49.run1 pat_3571          0.0   \n",
       "plt49.run1 pat_3572          0.0   \n",
       "\n",
       "                                                        rna parameters  \\\n",
       "rna project.run id                                                       \n",
       "plt1.run1 pat_88     mode: single-end, keep duplicates: no, keep mu...   \n",
       "plt1.run1 pat_89     mode: single-end, keep duplicates: no, keep mu...   \n",
       "...                                                                ...   \n",
       "plt49.run1 pat_3571                                                  0   \n",
       "plt49.run1 pat_3572                                                  0   \n",
       "\n",
       "                    rna reads processed rna reads with exactly 1 alignment  \\\n",
       "rna project.run id                                                           \n",
       "plt1.run1 pat_88             29,112,743                         23,152,813   \n",
       "plt1.run1 pat_89             24,355,220                         13,253,462   \n",
       "...                                 ...                                ...   \n",
       "plt49.run1 pat_3571                   0                                  0   \n",
       "plt49.run1 pat_3572                   0                                  0   \n",
       "\n",
       "                    rna reads with multiple alignments  \\\n",
       "rna project.run id                                       \n",
       "plt1.run1 pat_88                             4,044,454   \n",
       "plt1.run1 pat_89                             7,713,379   \n",
       "...                                                ...   \n",
       "plt49.run1 pat_3571                                  0   \n",
       "plt49.run1 pat_3572                                  0   \n",
       "\n",
       "                    rna reads with no alignment rna reads aligned  \\\n",
       "rna project.run id                                                  \n",
       "plt1.run1 pat_88                      1,915,476        27,197,267   \n",
       "plt1.run1 pat_89                      3,388,379        20,966,841   \n",
       "...                                         ...               ...   \n",
       "plt49.run1 pat_3571                           0                 0   \n",
       "plt49.run1 pat_3572                           0                 0   \n",
       "\n",
       "                     rna % reads aligned rna duplicate reads  \\\n",
       "rna project.run id                                             \n",
       "plt1.run1 pat_88                    93.0          20,093,139   \n",
       "plt1.run1 pat_89                    86.0          17,678,467   \n",
       "...                                  ...                 ...   \n",
       "plt49.run1 pat_3571                  0.0                   0   \n",
       "plt49.run1 pat_3572                  0.0                   0   \n",
       "\n",
       "                     rna % duplicate reads rna mitochondrial reads  \\\n",
       "rna project.run id                                                   \n",
       "plt1.run1 pat_88                      73.0               1,714,158   \n",
       "plt1.run1 pat_89                      84.0               1,475,002   \n",
       "...                                    ...                     ...   \n",
       "plt49.run1 pat_3571                    0.0                       0   \n",
       "plt49.run1 pat_3572                    0.0                       0   \n",
       "\n",
       "                     rna % mitochondrial reads rna rrna subunit reads  \\\n",
       "rna project.run id                                                      \n",
       "plt1.run1 pat_88                           6.0              1,868,907   \n",
       "plt1.run1 pat_89                           7.0              6,372,269   \n",
       "...                                        ...                    ...   \n",
       "plt49.run1 pat_3571                        0.0                      0   \n",
       "plt49.run1 pat_3572                        0.0                      0   \n",
       "\n",
       "                     rna % rrna subunit reads  rna pcr bottleneck coefficient  \\\n",
       "rna project.run id                                                              \n",
       "plt1.run1 pat_88                          6.0                            0.28   \n",
       "plt1.run1 pat_89                         30.0                            0.22   \n",
       "...                                       ...                             ...   \n",
       "plt49.run1 pat_3571                       0.0                            0.00   \n",
       "plt49.run1 pat_3572                       0.0                            0.00   \n",
       "\n",
       "                     rna mean fragment size (only for paired-end)  \\\n",
       "rna project.run id                                                  \n",
       "plt1.run1 pat_88                                              0.0   \n",
       "plt1.run1 pat_89                                              0.0   \n",
       "...                                                           ...   \n",
       "plt49.run1 pat_3571                                           0.0   \n",
       "plt49.run1 pat_3572                                           0.0   \n",
       "\n",
       "                    rna reads used for further steps (after optional filters for multimap, duplicate, mitochondria, rrna)  \\\n",
       "rna project.run id                                                                                                          \n",
       "plt1.run1 pat_88                                             6,644,895                                                      \n",
       "plt1.run1 pat_89                                             2,989,995                                                      \n",
       "...                                                                ...                                                      \n",
       "plt49.run1 pat_3571                                                  0                                                      \n",
       "plt49.run1 pat_3572                                                  0                                                      \n",
       "\n",
       "                                                      rna parameters.1  \\\n",
       "rna project.run id                                                       \n",
       "plt1.run1 pat_88     -t exon -g gene_id -s 2; multi-mapping: no; du...   \n",
       "plt1.run1 pat_89     -t exon -g gene_id -s 2; multi-mapping: no; du...   \n",
       "...                                                                ...   \n",
       "plt49.run1 pat_3571                                                  0   \n",
       "plt49.run1 pat_3572                                                  0   \n",
       "\n",
       "                    rna reads processed.1 rna reads assigned  \\\n",
       "rna project.run id                                             \n",
       "plt1.run1 pat_88                6,644,895          4,810,861   \n",
       "plt1.run1 pat_89                2,989,995          2,155,768   \n",
       "...                                   ...                ...   \n",
       "plt49.run1 pat_3571                     0                  0   \n",
       "plt49.run1 pat_3572                     0                  0   \n",
       "\n",
       "                     rna % reads assigned total  \\\n",
       "rna project.run id                                \n",
       "plt1.run1 pat_88                           16.0   \n",
       "plt1.run1 pat_89                            8.0   \n",
       "...                                         ...   \n",
       "plt49.run1 pat_3571                         0.0   \n",
       "plt49.run1 pat_3572                         0.0   \n",
       "\n",
       "                     rna % features with cpm >= 0.5  \\\n",
       "rna project.run id                                    \n",
       "plt1.run1 pat_88                               49.0   \n",
       "plt1.run1 pat_89                               45.0   \n",
       "...                                             ...   \n",
       "plt49.run1 pat_3571                             0.0   \n",
       "plt49.run1 pat_3572                             0.0   \n",
       "\n",
       "                     rna unassigned reads overlapping multiple genes  \\\n",
       "rna project.run id                                                     \n",
       "plt1.run1 pat_88                                             216.360   \n",
       "plt1.run1 pat_89                                              97.961   \n",
       "...                                                              ...   \n",
       "plt49.run1 pat_3571                                            0.000   \n",
       "plt49.run1 pat_3572                                            0.000   \n",
       "\n",
       "                     rna unassigned reads mapping to multiple locations  \\\n",
       "rna project.run id                                                        \n",
       "plt1.run1 pat_88                                                   0.0    \n",
       "plt1.run1 pat_89                                                   0.0    \n",
       "...                                                                ...    \n",
       "plt49.run1 pat_3571                                                0.0    \n",
       "plt49.run1 pat_3572                                                0.0    \n",
       "\n",
       "                    rna unassigned reads mapping to no feature  \\\n",
       "rna project.run id                                               \n",
       "plt1.run1 pat_88                                     1,617,674   \n",
       "plt1.run1 pat_89                                       736,266   \n",
       "...                                                        ...   \n",
       "plt49.run1 pat_3571                                          0   \n",
       "plt49.run1 pat_3572                                          0   \n",
       "\n",
       "                     rna unassigned reads not mapping  \\\n",
       "rna project.run id                                      \n",
       "plt1.run1 pat_88                                  0.0   \n",
       "plt1.run1 pat_89                                  0.0   \n",
       "...                                               ...   \n",
       "plt49.run1 pat_3571                               0.0   \n",
       "plt49.run1 pat_3572                               0.0   \n",
       "\n",
       "                     rna unassigned reads due to bad mapping quality  \\\n",
       "rna project.run id                                                     \n",
       "plt1.run1 pat_88                                                 0.0   \n",
       "plt1.run1 pat_89                                                 0.0   \n",
       "...                                                              ...   \n",
       "plt49.run1 pat_3571                                              0.0   \n",
       "plt49.run1 pat_3572                                              0.0   \n",
       "\n",
       "                     rna unassigned reads due to low fragment length  \\\n",
       "rna project.run id                                                     \n",
       "plt1.run1 pat_88                                                 0.0   \n",
       "plt1.run1 pat_89                                                 0.0   \n",
       "...                                                              ...   \n",
       "plt49.run1 pat_3571                                              0.0   \n",
       "plt49.run1 pat_3572                                              0.0   \n",
       "\n",
       "                     rna unassigned chimeric reads  \\\n",
       "rna project.run id                                   \n",
       "plt1.run1 pat_88                               0.0   \n",
       "plt1.run1 pat_89                               0.0   \n",
       "...                                            ...   \n",
       "plt49.run1 pat_3571                            0.0   \n",
       "plt49.run1 pat_3572                            0.0   \n",
       "\n",
       "                    rna usable sample: mapped reads > 2000000  rna DDX3Y  \\\n",
       "rna project.run id                                                         \n",
       "plt1.run1 pat_88                                         True      844.0   \n",
       "plt1.run1 pat_89                                         True      670.0   \n",
       "...                                                       ...        ...   \n",
       "plt49.run1 pat_3571                                         0        0.0   \n",
       "plt49.run1 pat_3572                                         0        0.0   \n",
       "\n",
       "                     rna XIST rna gender rna sample id original  \\\n",
       "rna project.run id                                                \n",
       "plt1.run1 pat_88          8.0          M              Sample 49   \n",
       "plt1.run1 pat_89          4.0          M              Sample 50   \n",
       "...                       ...        ...                    ...   \n",
       "plt49.run1 pat_3571       0.0          0                      0   \n",
       "plt49.run1 pat_3572       0.0          0                      0   \n",
       "\n",
       "                    rna plate id original    filename rel_path  \n",
       "rna project.run id                                              \n",
       "plt1.run1 pat_88                  Plate 1  matrix.mtx        .  \n",
       "plt1.run1 pat_89                  Plate 1  matrix.mtx        .  \n",
       "...                                   ...         ...      ...  \n",
       "plt49.run1 pat_3571                     0  matrix.mtx        .  \n",
       "plt49.run1 pat_3572                     0  matrix.mtx        .  \n",
       "\n",
       "[3661 rows x 76 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl gene</th>\n",
       "      <th>Ensembl biotype</th>\n",
       "      <th>UniProt proteins</th>\n",
       "      <th>UniProt genes</th>\n",
       "      <th>UniProt accessions</th>\n",
       "      <th>UniProt names</th>\n",
       "      <th>UniProt Ensembl transcripts</th>\n",
       "      <th>UniProt Ensembl proteins</th>\n",
       "      <th>UniProt Ensembl gene ids</th>\n",
       "      <th>Ensembl chr</th>\n",
       "      <th>Ensembl start</th>\n",
       "      <th>Ensembl stop</th>\n",
       "      <th>Ensembl strand</th>\n",
       "      <th>KEGG PATHWAY terms</th>\n",
       "      <th>KEGG PATHWAY ids</th>\n",
       "      <th>Gene Ontology terms</th>\n",
       "      <th>Gene Ontology ids</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ensembl gene id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000146555</th>\n",
       "      <td>SDK1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Protein sidekick-1</td>\n",
       "      <td>SDK1</td>\n",
       "      <td>Q7Z5N4; F8W6X9; A0A087WTQ6</td>\n",
       "      <td>SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN</td>\n",
       "      <td>ENST00000404826; ENST00000389531; ENST00000615806</td>\n",
       "      <td>ENSP00000385899; ENSP00000374182; ENSP00000478062</td>\n",
       "      <td>ENSG00000146555</td>\n",
       "      <td>chr7</td>\n",
       "      <td>3301252</td>\n",
       "      <td>4269000</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eye development, neural retina development, bi...</td>\n",
       "      <td>GO:0001654, GO:0003407, GO:0005488, GO:0005515...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000283537</th>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>chr7</td>\n",
       "      <td>143620943</td>\n",
       "      <td>143645675</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000270696</th>\n",
       "      <td>ENSG00000270696</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000270696</td>\n",
       "      <td>chr2</td>\n",
       "      <td>75660462</td>\n",
       "      <td>75662208</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000229906</th>\n",
       "      <td>SNRPGP11</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000229906</td>\n",
       "      <td>chr13</td>\n",
       "      <td>51496020</td>\n",
       "      <td>51496251</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57031 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Ensembl gene                     Ensembl biotype  \\\n",
       "Ensembl gene id                                                        \n",
       "ENSG00000146555             SDK1                      protein_coding   \n",
       "ENSG00000283537  ENSG00000283537  transcribed_unprocessed_pseudogene   \n",
       "...                          ...                                 ...   \n",
       "ENSG00000270696  ENSG00000270696                              lncRNA   \n",
       "ENSG00000229906         SNRPGP11                processed_pseudogene   \n",
       "\n",
       "                   UniProt proteins UniProt genes          UniProt accessions  \\\n",
       "Ensembl gene id                                                                 \n",
       "ENSG00000146555  Protein sidekick-1          SDK1  Q7Z5N4; F8W6X9; A0A087WTQ6   \n",
       "ENSG00000283537                 NaN           NaN                         NaN   \n",
       "...                             ...           ...                         ...   \n",
       "ENSG00000270696                 NaN           NaN                         NaN   \n",
       "ENSG00000229906                 NaN           NaN                         NaN   \n",
       "\n",
       "                                              UniProt names  \\\n",
       "Ensembl gene id                                               \n",
       "ENSG00000146555  SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN   \n",
       "ENSG00000283537                                         NaN   \n",
       "...                                                     ...   \n",
       "ENSG00000270696                                         NaN   \n",
       "ENSG00000229906                                         NaN   \n",
       "\n",
       "                                       UniProt Ensembl transcripts  \\\n",
       "Ensembl gene id                                                      \n",
       "ENSG00000146555  ENST00000404826; ENST00000389531; ENST00000615806   \n",
       "ENSG00000283537                                                NaN   \n",
       "...                                                            ...   \n",
       "ENSG00000270696                                                NaN   \n",
       "ENSG00000229906                                                NaN   \n",
       "\n",
       "                                          UniProt Ensembl proteins  \\\n",
       "Ensembl gene id                                                      \n",
       "ENSG00000146555  ENSP00000385899; ENSP00000374182; ENSP00000478062   \n",
       "ENSG00000283537                                                NaN   \n",
       "...                                                            ...   \n",
       "ENSG00000270696                                                NaN   \n",
       "ENSG00000229906                                                NaN   \n",
       "\n",
       "                UniProt Ensembl gene ids Ensembl chr  Ensembl start  \\\n",
       "Ensembl gene id                                                       \n",
       "ENSG00000146555          ENSG00000146555        chr7        3301252   \n",
       "ENSG00000283537          ENSG00000283537        chr7      143620943   \n",
       "...                                  ...         ...            ...   \n",
       "ENSG00000270696          ENSG00000270696        chr2       75660462   \n",
       "ENSG00000229906          ENSG00000229906       chr13       51496020   \n",
       "\n",
       "                 Ensembl stop Ensembl strand KEGG PATHWAY terms  \\\n",
       "Ensembl gene id                                                   \n",
       "ENSG00000146555       4269000              +                NaN   \n",
       "ENSG00000283537     143645675              +                NaN   \n",
       "...                       ...            ...                ...   \n",
       "ENSG00000270696      75662208              +                NaN   \n",
       "ENSG00000229906      51496251              -                NaN   \n",
       "\n",
       "                KEGG PATHWAY ids  \\\n",
       "Ensembl gene id                    \n",
       "ENSG00000146555              NaN   \n",
       "ENSG00000283537              NaN   \n",
       "...                          ...   \n",
       "ENSG00000270696              NaN   \n",
       "ENSG00000229906              NaN   \n",
       "\n",
       "                                               Gene Ontology terms  \\\n",
       "Ensembl gene id                                                      \n",
       "ENSG00000146555  eye development, neural retina development, bi...   \n",
       "ENSG00000283537                                                NaN   \n",
       "...                                                            ...   \n",
       "ENSG00000270696                                                NaN   \n",
       "ENSG00000229906                                                NaN   \n",
       "\n",
       "                                                 Gene Ontology ids  \n",
       "Ensembl gene id                                                     \n",
       "ENSG00000146555  GO:0001654, GO:0003407, GO:0005488, GO:0005515...  \n",
       "ENSG00000283537                                                NaN  \n",
       "...                                                            ...  \n",
       "ENSG00000270696                                                NaN  \n",
       "ENSG00000229906                                                NaN  \n",
       "\n",
       "[57031 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context('display.max_rows', 5,'display.max_columns', None):\n",
    "    display(adata.obs)\n",
    "    display(adata.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TubeID',\n",
       " 'meta TubeID',\n",
       " 'meta pop export_psn',\n",
       " 'meta pop visit_1_date',\n",
       " 'meta pop visit_2_date',\n",
       " 'meta pop visit_3_date',\n",
       " 'meta pop visit_4_date',\n",
       " 'meta pop age',\n",
       " 'meta pop age_dec',\n",
       " 'meta pop sex',\n",
       " 'meta pop c19_severity',\n",
       " 'meta pop cohort',\n",
       " 'meta suep_hap export_psn',\n",
       " 'meta suep_hap baseline_date',\n",
       " 'meta suep_hap end_acute_date',\n",
       " 'meta suep_hap end_acute_visit_date',\n",
       " 'meta suep_hap m3_fu_date',\n",
       " 'meta suep_hap m6_fu_date',\n",
       " 'meta suep_hap m12_fu_date',\n",
       " 'meta suep_hap m24_fu_date',\n",
       " 'meta suep_hap age',\n",
       " 'meta suep_hap age_dec',\n",
       " 'meta suep_hap sex',\n",
       " 'meta suep_hap c19_severity',\n",
       " 'meta suep_hap cohort',\n",
       " 'meta export_psn',\n",
       " 'meta age',\n",
       " 'meta age_dec',\n",
       " 'meta sex',\n",
       " 'meta c19_severity',\n",
       " 'meta cohort',\n",
       " 'rna TubeID',\n",
       " 'rna id',\n",
       " 'rna fastq1',\n",
       " 'rna project',\n",
       " 'rna run',\n",
       " 'rna plate',\n",
       " 'rna organism',\n",
       " 'rna assembly',\n",
       " 'rna release',\n",
       " 'rna parameters',\n",
       " 'rna reads processed',\n",
       " 'rna reads with exactly 1 alignment',\n",
       " 'rna reads with multiple alignments',\n",
       " 'rna reads with no alignment',\n",
       " 'rna reads aligned',\n",
       " 'rna % reads aligned',\n",
       " 'rna duplicate reads',\n",
       " 'rna % duplicate reads',\n",
       " 'rna mitochondrial reads',\n",
       " 'rna % mitochondrial reads',\n",
       " 'rna rrna subunit reads',\n",
       " 'rna % rrna subunit reads',\n",
       " 'rna pcr bottleneck coefficient',\n",
       " 'rna mean fragment size (only for paired-end)',\n",
       " 'rna reads used for further steps (after optional filters for multimap, duplicate, mitochondria, rrna)',\n",
       " 'rna parameters.1',\n",
       " 'rna reads processed.1',\n",
       " 'rna reads assigned',\n",
       " 'rna % reads assigned total',\n",
       " 'rna % features with cpm >= 0.5',\n",
       " 'rna unassigned reads overlapping multiple genes',\n",
       " 'rna unassigned reads mapping to multiple locations',\n",
       " 'rna unassigned reads mapping to no feature',\n",
       " 'rna unassigned reads not mapping',\n",
       " 'rna unassigned reads due to bad mapping quality',\n",
       " 'rna unassigned reads due to low fragment length',\n",
       " 'rna unassigned chimeric reads',\n",
       " 'rna usable sample: mapped reads > 2000000',\n",
       " 'rna project.run id',\n",
       " 'rna DDX3Y',\n",
       " 'rna XIST',\n",
       " 'rna gender',\n",
       " 'rna sample id original',\n",
       " 'rna plate id original']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pat_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['rna usable sample: mapped reads > 2000000'] = adata.obs['rna usable sample: mapped reads > 2000000'].astype(str) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# .obs-Spaltennamen, die gelöscht werden sollen\n",
    "drop_obs = []\n",
    "\n",
    "# .obs-Spaltennamen, die geändert werden sollen. z.B. \"old_name\": \"new_name\"\n",
    "# Beispiel: \"old_name\" = \"rna usable sample: mapped reads > 2000000\" -> Regex bevorzugt\n",
    "rename_obs = {'rna reads used for further steps (after optional filters for multimap, duplicate, mitochondria, rrna)':'rna reads used after filters', 'rna usable sample: mapped reads > 2000000':'mapped reads > 2000000'} \n",
    "\n",
    "#Namen ändern mithilfe von Regex\n",
    "replace_obs = {'unassigned':'unass.','due to':'',' ':'_','%':'pct',':':'_of','>=':'at_least','>':'over'}\n",
    "replace_var = {' ':'_'}\n",
    "\n",
    "# .var-Spaltennamen, die gelöscht werden sollen\n",
    "drop_var = []\n",
    "\n",
    "# .var-Spaltennamen, die geändert werden sollen. z.B. \"old_name\": \"new_name\"\n",
    "rename_var = {}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Änderungen an .obs-Spaltennamen in eine Kopie absspeichern\n",
    "obs = adata.obs.copy()\n",
    "\n",
    "obs.drop(columns=drop_obs, inplace=True)\n",
    "obs.rename(columns=rename_obs, errors='raise', inplace=True)\n",
    "\n",
    "#replace\n",
    "for x, y in replace_obs.items():\n",
    "    obs.columns= obs.columns.str.replace(x, y,regex=True)\n",
    "    \n",
    "    \n",
    "# Änderungen an .var-Spaltennamen in eine Kopie absspeichern\n",
    "var = adata.var.copy()\n",
    "\n",
    "var.drop(columns=drop_var, inplace=True)\n",
    "var.rename(columns=rename_var, errors='raise', inplace=True)\n",
    "\n",
    "for x, y in replace_var.items():\n",
    "    var.columns= var.columns.str.replace(x, y,regex=True) \n",
    "\n",
    "# Änderungen an AnnData anwenden\n",
    "adata.obs = obs\n",
    "adata.var = var\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7. --- AnnData-Objekt speichern ---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3661 × 57031\n",
       "    obs: 'TubeID', 'meta_TubeID', 'meta_pop_export_psn', 'meta_pop_visit_1_date', 'meta_pop_visit_2_date', 'meta_pop_visit_3_date', 'meta_pop_visit_4_date', 'meta_pop_age', 'meta_pop_age_dec', 'meta_pop_sex', 'meta_pop_c19_severity', 'meta_pop_cohort', 'meta_suep_hap_export_psn', 'meta_suep_hap_baseline_date', 'meta_suep_hap_end_acute_date', 'meta_suep_hap_end_acute_visit_date', 'meta_suep_hap_m3_fu_date', 'meta_suep_hap_m6_fu_date', 'meta_suep_hap_m12_fu_date', 'meta_suep_hap_m24_fu_date', 'meta_suep_hap_age', 'meta_suep_hap_age_dec', 'meta_suep_hap_sex', 'meta_suep_hap_c19_severity', 'meta_suep_hap_cohort', 'meta_export_psn', 'meta_age', 'meta_age_dec', 'meta_sex', 'meta_c19_severity', 'meta_cohort', 'rna_TubeID', 'rna_id', 'rna_fastq1', 'rna_project', 'rna_run', 'rna_plate', 'rna_organism', 'rna_assembly', 'rna_release', 'rna_parameters', 'rna_reads_processed', 'rna_reads_with_exactly_1_alignment', 'rna_reads_with_multiple_alignments', 'rna_reads_with_no_alignment', 'rna_reads_aligned', 'rna_pct_reads_aligned', 'rna_duplicate_reads', 'rna_pct_duplicate_reads', 'rna_mitochondrial_reads', 'rna_pct_mitochondrial_reads', 'rna_rrna_subunit_reads', 'rna_pct_rrna_subunit_reads', 'rna_pcr_bottleneck_coefficient', 'rna_mean_fragment_size_(only_for_paired-end)', 'rna_reads_used_after_filters', 'rna_parameters.1', 'rna_reads_processed.1', 'rna_reads_assigned', 'rna_pct_reads_assigned_total', 'rna_pct_features_with_cpm_at_least_0.5', 'rna_unass._reads_overlapping_multiple_genes', 'rna_unass._reads_mapping_to_multiple_locations', 'rna_unass._reads_mapping_to_no_feature', 'rna_unass._reads_not_mapping', 'rna_unass._reads__bad_mapping_quality', 'rna_unass._reads__low_fragment_length', 'rna_unass._chimeric_reads', 'mapped_reads_over_2000000', 'rna_DDX3Y', 'rna_XIST', 'rna_gender', 'rna_sample_id_original', 'rna_plate_id_original', 'filename', 'rel_path'\n",
       "    var: 'Ensembl_gene', 'Ensembl_biotype', 'UniProt_proteins', 'UniProt_genes', 'UniProt_accessions', 'UniProt_names', 'UniProt_Ensembl_transcripts', 'UniProt_Ensembl_proteins', 'UniProt_Ensembl_gene_ids', 'Ensembl_chr', 'Ensembl_start', 'Ensembl_stop', 'Ensembl_strand', 'KEGG_PATHWAY_terms', 'KEGG_PATHWAY_ids', 'Gene_Ontology_terms', 'Gene_Ontology_ids'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TubeID</th>\n",
       "      <th>meta_TubeID</th>\n",
       "      <th>meta_pop_export_psn</th>\n",
       "      <th>meta_pop_visit_1_date</th>\n",
       "      <th>meta_pop_visit_2_date</th>\n",
       "      <th>meta_pop_visit_3_date</th>\n",
       "      <th>meta_pop_visit_4_date</th>\n",
       "      <th>meta_pop_age</th>\n",
       "      <th>meta_pop_age_dec</th>\n",
       "      <th>meta_pop_sex</th>\n",
       "      <th>meta_pop_c19_severity</th>\n",
       "      <th>meta_pop_cohort</th>\n",
       "      <th>meta_suep_hap_export_psn</th>\n",
       "      <th>meta_suep_hap_baseline_date</th>\n",
       "      <th>meta_suep_hap_end_acute_date</th>\n",
       "      <th>meta_suep_hap_end_acute_visit_date</th>\n",
       "      <th>meta_suep_hap_m3_fu_date</th>\n",
       "      <th>meta_suep_hap_m6_fu_date</th>\n",
       "      <th>meta_suep_hap_m12_fu_date</th>\n",
       "      <th>meta_suep_hap_m24_fu_date</th>\n",
       "      <th>meta_suep_hap_age</th>\n",
       "      <th>meta_suep_hap_age_dec</th>\n",
       "      <th>meta_suep_hap_sex</th>\n",
       "      <th>meta_suep_hap_c19_severity</th>\n",
       "      <th>meta_suep_hap_cohort</th>\n",
       "      <th>meta_export_psn</th>\n",
       "      <th>meta_age</th>\n",
       "      <th>meta_age_dec</th>\n",
       "      <th>meta_sex</th>\n",
       "      <th>meta_c19_severity</th>\n",
       "      <th>meta_cohort</th>\n",
       "      <th>rna_TubeID</th>\n",
       "      <th>rna_id</th>\n",
       "      <th>rna_fastq1</th>\n",
       "      <th>rna_project</th>\n",
       "      <th>rna_run</th>\n",
       "      <th>rna_plate</th>\n",
       "      <th>rna_organism</th>\n",
       "      <th>rna_assembly</th>\n",
       "      <th>rna_release</th>\n",
       "      <th>rna_parameters</th>\n",
       "      <th>rna_reads_processed</th>\n",
       "      <th>rna_reads_with_exactly_1_alignment</th>\n",
       "      <th>rna_reads_with_multiple_alignments</th>\n",
       "      <th>rna_reads_with_no_alignment</th>\n",
       "      <th>rna_reads_aligned</th>\n",
       "      <th>rna_pct_reads_aligned</th>\n",
       "      <th>rna_duplicate_reads</th>\n",
       "      <th>rna_pct_duplicate_reads</th>\n",
       "      <th>rna_mitochondrial_reads</th>\n",
       "      <th>rna_pct_mitochondrial_reads</th>\n",
       "      <th>rna_rrna_subunit_reads</th>\n",
       "      <th>rna_pct_rrna_subunit_reads</th>\n",
       "      <th>rna_pcr_bottleneck_coefficient</th>\n",
       "      <th>rna_mean_fragment_size_(only_for_paired-end)</th>\n",
       "      <th>rna_reads_used_after_filters</th>\n",
       "      <th>rna_parameters.1</th>\n",
       "      <th>rna_reads_processed.1</th>\n",
       "      <th>rna_reads_assigned</th>\n",
       "      <th>rna_pct_reads_assigned_total</th>\n",
       "      <th>rna_pct_features_with_cpm_at_least_0.5</th>\n",
       "      <th>rna_unass._reads_overlapping_multiple_genes</th>\n",
       "      <th>rna_unass._reads_mapping_to_multiple_locations</th>\n",
       "      <th>rna_unass._reads_mapping_to_no_feature</th>\n",
       "      <th>rna_unass._reads_not_mapping</th>\n",
       "      <th>rna_unass._reads__bad_mapping_quality</th>\n",
       "      <th>rna_unass._reads__low_fragment_length</th>\n",
       "      <th>rna_unass._chimeric_reads</th>\n",
       "      <th>mapped_reads_over_2000000</th>\n",
       "      <th>rna_DDX3Y</th>\n",
       "      <th>rna_XIST</th>\n",
       "      <th>rna_gender</th>\n",
       "      <th>rna_sample_id_original</th>\n",
       "      <th>rna_plate_id_original</th>\n",
       "      <th>filename</th>\n",
       "      <th>rel_path</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rna project.run id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>plt1.run1 pat_88</th>\n",
       "      <td>LV1007699778</td>\n",
       "      <td>LV1007699778</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>hap_0178</td>\n",
       "      <td>17/11/2021</td>\n",
       "      <td>24/11/2021</td>\n",
       "      <td>24/11/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>60-69</td>\n",
       "      <td>M</td>\n",
       "      <td>severe</td>\n",
       "      <td>hap</td>\n",
       "      <td>hap_0178</td>\n",
       "      <td>65.0</td>\n",
       "      <td>60-69</td>\n",
       "      <td>M</td>\n",
       "      <td>severe</td>\n",
       "      <td>hap</td>\n",
       "      <td>LV1007699778</td>\n",
       "      <td>pat_88</td>\n",
       "      <td>AAAVY52HV_NAPKON_Plate1_2_22s004137-1-1_Valasa...</td>\n",
       "      <td>plt1</td>\n",
       "      <td>run1</td>\n",
       "      <td>plate1</td>\n",
       "      <td>human</td>\n",
       "      <td>hg38</td>\n",
       "      <td>104.0</td>\n",
       "      <td>mode: single-end, keep duplicates: no, keep mu...</td>\n",
       "      <td>29,112,743</td>\n",
       "      <td>23,152,813</td>\n",
       "      <td>4,044,454</td>\n",
       "      <td>1,915,476</td>\n",
       "      <td>27,197,267</td>\n",
       "      <td>93.0</td>\n",
       "      <td>20,093,139</td>\n",
       "      <td>73.0</td>\n",
       "      <td>1,714,158</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1,868,907</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6,644,895</td>\n",
       "      <td>-t exon -g gene_id -s 2; multi-mapping: no; du...</td>\n",
       "      <td>6,644,895</td>\n",
       "      <td>4,810,861</td>\n",
       "      <td>16.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>216.360</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1,617,674</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>844.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Sample 49</td>\n",
       "      <td>Plate 1</td>\n",
       "      <td>matrix.mtx</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plt1.run1 pat_89</th>\n",
       "      <td>LV1007699743</td>\n",
       "      <td>LV1007699743</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>suep_0425</td>\n",
       "      <td>08/11/2021</td>\n",
       "      <td>17/11/2021</td>\n",
       "      <td>17/11/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>60-69</td>\n",
       "      <td>M</td>\n",
       "      <td>mild</td>\n",
       "      <td>suep</td>\n",
       "      <td>suep_0425</td>\n",
       "      <td>61.0</td>\n",
       "      <td>60-69</td>\n",
       "      <td>M</td>\n",
       "      <td>mild</td>\n",
       "      <td>suep</td>\n",
       "      <td>LV1007699743</td>\n",
       "      <td>pat_89</td>\n",
       "      <td>AAAVY52HV_NAPKON_Plate1_2_22s004137-1-1_Valasa...</td>\n",
       "      <td>plt1</td>\n",
       "      <td>run1</td>\n",
       "      <td>plate1</td>\n",
       "      <td>human</td>\n",
       "      <td>hg38</td>\n",
       "      <td>104.0</td>\n",
       "      <td>mode: single-end, keep duplicates: no, keep mu...</td>\n",
       "      <td>24,355,220</td>\n",
       "      <td>13,253,462</td>\n",
       "      <td>7,713,379</td>\n",
       "      <td>3,388,379</td>\n",
       "      <td>20,966,841</td>\n",
       "      <td>86.0</td>\n",
       "      <td>17,678,467</td>\n",
       "      <td>84.0</td>\n",
       "      <td>1,475,002</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6,372,269</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2,989,995</td>\n",
       "      <td>-t exon -g gene_id -s 2; multi-mapping: no; du...</td>\n",
       "      <td>2,989,995</td>\n",
       "      <td>2,155,768</td>\n",
       "      <td>8.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>97.961</td>\n",
       "      <td>0.0</td>\n",
       "      <td>736,266</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>670.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Sample 50</td>\n",
       "      <td>Plate 1</td>\n",
       "      <td>matrix.mtx</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plt49.run1 pat_3571</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>matrix.mtx</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plt49.run1 pat_3572</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>matrix.mtx</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3661 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           TubeID   meta_TubeID meta_pop_export_psn  \\\n",
       "rna project.run id                                                    \n",
       "plt1.run1 pat_88     LV1007699778  LV1007699778                   0   \n",
       "plt1.run1 pat_89     LV1007699743  LV1007699743                   0   \n",
       "...                           ...           ...                 ...   \n",
       "plt49.run1 pat_3571             0             0                   0   \n",
       "plt49.run1 pat_3572             0             0                   0   \n",
       "\n",
       "                    meta_pop_visit_1_date meta_pop_visit_2_date  \\\n",
       "rna project.run id                                                \n",
       "plt1.run1 pat_88                        0                     0   \n",
       "plt1.run1 pat_89                        0                     0   \n",
       "...                                   ...                   ...   \n",
       "plt49.run1 pat_3571                     0                     0   \n",
       "plt49.run1 pat_3572                     0                     0   \n",
       "\n",
       "                    meta_pop_visit_3_date meta_pop_visit_4_date  meta_pop_age  \\\n",
       "rna project.run id                                                              \n",
       "plt1.run1 pat_88                        0                     0           0.0   \n",
       "plt1.run1 pat_89                        0                     0           0.0   \n",
       "...                                   ...                   ...           ...   \n",
       "plt49.run1 pat_3571                     0                     0           0.0   \n",
       "plt49.run1 pat_3572                     0                     0           0.0   \n",
       "\n",
       "                    meta_pop_age_dec meta_pop_sex meta_pop_c19_severity  \\\n",
       "rna project.run id                                                        \n",
       "plt1.run1 pat_88                   0            0                     0   \n",
       "plt1.run1 pat_89                   0            0                     0   \n",
       "...                              ...          ...                   ...   \n",
       "plt49.run1 pat_3571                0            0                     0   \n",
       "plt49.run1 pat_3572                0            0                     0   \n",
       "\n",
       "                    meta_pop_cohort meta_suep_hap_export_psn  \\\n",
       "rna project.run id                                             \n",
       "plt1.run1 pat_88                  0                 hap_0178   \n",
       "plt1.run1 pat_89                  0                suep_0425   \n",
       "...                             ...                      ...   \n",
       "plt49.run1 pat_3571               0                        0   \n",
       "plt49.run1 pat_3572               0                        0   \n",
       "\n",
       "                    meta_suep_hap_baseline_date meta_suep_hap_end_acute_date  \\\n",
       "rna project.run id                                                             \n",
       "plt1.run1 pat_88                     17/11/2021                   24/11/2021   \n",
       "plt1.run1 pat_89                     08/11/2021                   17/11/2021   \n",
       "...                                         ...                          ...   \n",
       "plt49.run1 pat_3571                           0                            0   \n",
       "plt49.run1 pat_3572                           0                            0   \n",
       "\n",
       "                    meta_suep_hap_end_acute_visit_date  \\\n",
       "rna project.run id                                       \n",
       "plt1.run1 pat_88                            24/11/2021   \n",
       "plt1.run1 pat_89                            17/11/2021   \n",
       "...                                                ...   \n",
       "plt49.run1 pat_3571                                  0   \n",
       "plt49.run1 pat_3572                                  0   \n",
       "\n",
       "                    meta_suep_hap_m3_fu_date meta_suep_hap_m6_fu_date  \\\n",
       "rna project.run id                                                      \n",
       "plt1.run1 pat_88                           0                        0   \n",
       "plt1.run1 pat_89                           0                        0   \n",
       "...                                      ...                      ...   \n",
       "plt49.run1 pat_3571                        0                        0   \n",
       "plt49.run1 pat_3572                        0                        0   \n",
       "\n",
       "                    meta_suep_hap_m12_fu_date meta_suep_hap_m24_fu_date  \\\n",
       "rna project.run id                                                        \n",
       "plt1.run1 pat_88                            0                         0   \n",
       "plt1.run1 pat_89                            0                         0   \n",
       "...                                       ...                       ...   \n",
       "plt49.run1 pat_3571                         0                         0   \n",
       "plt49.run1 pat_3572                         0                         0   \n",
       "\n",
       "                     meta_suep_hap_age meta_suep_hap_age_dec  \\\n",
       "rna project.run id                                             \n",
       "plt1.run1 pat_88                  65.0                 60-69   \n",
       "plt1.run1 pat_89                  61.0                 60-69   \n",
       "...                                ...                   ...   \n",
       "plt49.run1 pat_3571                0.0                     0   \n",
       "plt49.run1 pat_3572                0.0                     0   \n",
       "\n",
       "                    meta_suep_hap_sex meta_suep_hap_c19_severity  \\\n",
       "rna project.run id                                                 \n",
       "plt1.run1 pat_88                    M                     severe   \n",
       "plt1.run1 pat_89                    M                       mild   \n",
       "...                               ...                        ...   \n",
       "plt49.run1 pat_3571                 0                          0   \n",
       "plt49.run1 pat_3572                 0                          0   \n",
       "\n",
       "                    meta_suep_hap_cohort meta_export_psn  meta_age  \\\n",
       "rna project.run id                                                   \n",
       "plt1.run1 pat_88                     hap        hap_0178      65.0   \n",
       "plt1.run1 pat_89                    suep       suep_0425      61.0   \n",
       "...                                  ...             ...       ...   \n",
       "plt49.run1 pat_3571                    0               0       0.0   \n",
       "plt49.run1 pat_3572                    0               0       0.0   \n",
       "\n",
       "                    meta_age_dec meta_sex meta_c19_severity meta_cohort  \\\n",
       "rna project.run id                                                        \n",
       "plt1.run1 pat_88           60-69        M            severe         hap   \n",
       "plt1.run1 pat_89           60-69        M              mild        suep   \n",
       "...                          ...      ...               ...         ...   \n",
       "plt49.run1 pat_3571            0        0                 0           0   \n",
       "plt49.run1 pat_3572            0        0                 0           0   \n",
       "\n",
       "                       rna_TubeID  rna_id  \\\n",
       "rna project.run id                          \n",
       "plt1.run1 pat_88     LV1007699778  pat_88   \n",
       "plt1.run1 pat_89     LV1007699743  pat_89   \n",
       "...                           ...     ...   \n",
       "plt49.run1 pat_3571             0       0   \n",
       "plt49.run1 pat_3572             0       0   \n",
       "\n",
       "                                                            rna_fastq1  \\\n",
       "rna project.run id                                                       \n",
       "plt1.run1 pat_88     AAAVY52HV_NAPKON_Plate1_2_22s004137-1-1_Valasa...   \n",
       "plt1.run1 pat_89     AAAVY52HV_NAPKON_Plate1_2_22s004137-1-1_Valasa...   \n",
       "...                                                                ...   \n",
       "plt49.run1 pat_3571                                                  0   \n",
       "plt49.run1 pat_3572                                                  0   \n",
       "\n",
       "                    rna_project rna_run rna_plate rna_organism rna_assembly  \\\n",
       "rna project.run id                                                            \n",
       "plt1.run1 pat_88           plt1    run1    plate1        human         hg38   \n",
       "plt1.run1 pat_89           plt1    run1    plate1        human         hg38   \n",
       "...                         ...     ...       ...          ...          ...   \n",
       "plt49.run1 pat_3571           0       0         0            0            0   \n",
       "plt49.run1 pat_3572           0       0         0            0            0   \n",
       "\n",
       "                     rna_release  \\\n",
       "rna project.run id                 \n",
       "plt1.run1 pat_88           104.0   \n",
       "plt1.run1 pat_89           104.0   \n",
       "...                          ...   \n",
       "plt49.run1 pat_3571          0.0   \n",
       "plt49.run1 pat_3572          0.0   \n",
       "\n",
       "                                                        rna_parameters  \\\n",
       "rna project.run id                                                       \n",
       "plt1.run1 pat_88     mode: single-end, keep duplicates: no, keep mu...   \n",
       "plt1.run1 pat_89     mode: single-end, keep duplicates: no, keep mu...   \n",
       "...                                                                ...   \n",
       "plt49.run1 pat_3571                                                  0   \n",
       "plt49.run1 pat_3572                                                  0   \n",
       "\n",
       "                    rna_reads_processed rna_reads_with_exactly_1_alignment  \\\n",
       "rna project.run id                                                           \n",
       "plt1.run1 pat_88             29,112,743                         23,152,813   \n",
       "plt1.run1 pat_89             24,355,220                         13,253,462   \n",
       "...                                 ...                                ...   \n",
       "plt49.run1 pat_3571                   0                                  0   \n",
       "plt49.run1 pat_3572                   0                                  0   \n",
       "\n",
       "                    rna_reads_with_multiple_alignments  \\\n",
       "rna project.run id                                       \n",
       "plt1.run1 pat_88                             4,044,454   \n",
       "plt1.run1 pat_89                             7,713,379   \n",
       "...                                                ...   \n",
       "plt49.run1 pat_3571                                  0   \n",
       "plt49.run1 pat_3572                                  0   \n",
       "\n",
       "                    rna_reads_with_no_alignment rna_reads_aligned  \\\n",
       "rna project.run id                                                  \n",
       "plt1.run1 pat_88                      1,915,476        27,197,267   \n",
       "plt1.run1 pat_89                      3,388,379        20,966,841   \n",
       "...                                         ...               ...   \n",
       "plt49.run1 pat_3571                           0                 0   \n",
       "plt49.run1 pat_3572                           0                 0   \n",
       "\n",
       "                     rna_pct_reads_aligned rna_duplicate_reads  \\\n",
       "rna project.run id                                               \n",
       "plt1.run1 pat_88                      93.0          20,093,139   \n",
       "plt1.run1 pat_89                      86.0          17,678,467   \n",
       "...                                    ...                 ...   \n",
       "plt49.run1 pat_3571                    0.0                   0   \n",
       "plt49.run1 pat_3572                    0.0                   0   \n",
       "\n",
       "                     rna_pct_duplicate_reads rna_mitochondrial_reads  \\\n",
       "rna project.run id                                                     \n",
       "plt1.run1 pat_88                        73.0               1,714,158   \n",
       "plt1.run1 pat_89                        84.0               1,475,002   \n",
       "...                                      ...                     ...   \n",
       "plt49.run1 pat_3571                      0.0                       0   \n",
       "plt49.run1 pat_3572                      0.0                       0   \n",
       "\n",
       "                     rna_pct_mitochondrial_reads rna_rrna_subunit_reads  \\\n",
       "rna project.run id                                                        \n",
       "plt1.run1 pat_88                             6.0              1,868,907   \n",
       "plt1.run1 pat_89                             7.0              6,372,269   \n",
       "...                                          ...                    ...   \n",
       "plt49.run1 pat_3571                          0.0                      0   \n",
       "plt49.run1 pat_3572                          0.0                      0   \n",
       "\n",
       "                     rna_pct_rrna_subunit_reads  \\\n",
       "rna project.run id                                \n",
       "plt1.run1 pat_88                            6.0   \n",
       "plt1.run1 pat_89                           30.0   \n",
       "...                                         ...   \n",
       "plt49.run1 pat_3571                         0.0   \n",
       "plt49.run1 pat_3572                         0.0   \n",
       "\n",
       "                     rna_pcr_bottleneck_coefficient  \\\n",
       "rna project.run id                                    \n",
       "plt1.run1 pat_88                               0.28   \n",
       "plt1.run1 pat_89                               0.22   \n",
       "...                                             ...   \n",
       "plt49.run1 pat_3571                            0.00   \n",
       "plt49.run1 pat_3572                            0.00   \n",
       "\n",
       "                     rna_mean_fragment_size_(only_for_paired-end)  \\\n",
       "rna project.run id                                                  \n",
       "plt1.run1 pat_88                                              0.0   \n",
       "plt1.run1 pat_89                                              0.0   \n",
       "...                                                           ...   \n",
       "plt49.run1 pat_3571                                           0.0   \n",
       "plt49.run1 pat_3572                                           0.0   \n",
       "\n",
       "                    rna_reads_used_after_filters  \\\n",
       "rna project.run id                                 \n",
       "plt1.run1 pat_88                       6,644,895   \n",
       "plt1.run1 pat_89                       2,989,995   \n",
       "...                                          ...   \n",
       "plt49.run1 pat_3571                            0   \n",
       "plt49.run1 pat_3572                            0   \n",
       "\n",
       "                                                      rna_parameters.1  \\\n",
       "rna project.run id                                                       \n",
       "plt1.run1 pat_88     -t exon -g gene_id -s 2; multi-mapping: no; du...   \n",
       "plt1.run1 pat_89     -t exon -g gene_id -s 2; multi-mapping: no; du...   \n",
       "...                                                                ...   \n",
       "plt49.run1 pat_3571                                                  0   \n",
       "plt49.run1 pat_3572                                                  0   \n",
       "\n",
       "                    rna_reads_processed.1 rna_reads_assigned  \\\n",
       "rna project.run id                                             \n",
       "plt1.run1 pat_88                6,644,895          4,810,861   \n",
       "plt1.run1 pat_89                2,989,995          2,155,768   \n",
       "...                                   ...                ...   \n",
       "plt49.run1 pat_3571                     0                  0   \n",
       "plt49.run1 pat_3572                     0                  0   \n",
       "\n",
       "                     rna_pct_reads_assigned_total  \\\n",
       "rna project.run id                                  \n",
       "plt1.run1 pat_88                             16.0   \n",
       "plt1.run1 pat_89                              8.0   \n",
       "...                                           ...   \n",
       "plt49.run1 pat_3571                           0.0   \n",
       "plt49.run1 pat_3572                           0.0   \n",
       "\n",
       "                     rna_pct_features_with_cpm_at_least_0.5  \\\n",
       "rna project.run id                                            \n",
       "plt1.run1 pat_88                                       49.0   \n",
       "plt1.run1 pat_89                                       45.0   \n",
       "...                                                     ...   \n",
       "plt49.run1 pat_3571                                     0.0   \n",
       "plt49.run1 pat_3572                                     0.0   \n",
       "\n",
       "                     rna_unass._reads_overlapping_multiple_genes  \\\n",
       "rna project.run id                                                 \n",
       "plt1.run1 pat_88                                         216.360   \n",
       "plt1.run1 pat_89                                          97.961   \n",
       "...                                                          ...   \n",
       "plt49.run1 pat_3571                                        0.000   \n",
       "plt49.run1 pat_3572                                        0.000   \n",
       "\n",
       "                     rna_unass._reads_mapping_to_multiple_locations  \\\n",
       "rna project.run id                                                    \n",
       "plt1.run1 pat_88                                                0.0   \n",
       "plt1.run1 pat_89                                                0.0   \n",
       "...                                                             ...   \n",
       "plt49.run1 pat_3571                                             0.0   \n",
       "plt49.run1 pat_3572                                             0.0   \n",
       "\n",
       "                    rna_unass._reads_mapping_to_no_feature  \\\n",
       "rna project.run id                                           \n",
       "plt1.run1 pat_88                                 1,617,674   \n",
       "plt1.run1 pat_89                                   736,266   \n",
       "...                                                    ...   \n",
       "plt49.run1 pat_3571                                      0   \n",
       "plt49.run1 pat_3572                                      0   \n",
       "\n",
       "                     rna_unass._reads_not_mapping  \\\n",
       "rna project.run id                                  \n",
       "plt1.run1 pat_88                              0.0   \n",
       "plt1.run1 pat_89                              0.0   \n",
       "...                                           ...   \n",
       "plt49.run1 pat_3571                           0.0   \n",
       "plt49.run1 pat_3572                           0.0   \n",
       "\n",
       "                     rna_unass._reads__bad_mapping_quality  \\\n",
       "rna project.run id                                           \n",
       "plt1.run1 pat_88                                       0.0   \n",
       "plt1.run1 pat_89                                       0.0   \n",
       "...                                                    ...   \n",
       "plt49.run1 pat_3571                                    0.0   \n",
       "plt49.run1 pat_3572                                    0.0   \n",
       "\n",
       "                     rna_unass._reads__low_fragment_length  \\\n",
       "rna project.run id                                           \n",
       "plt1.run1 pat_88                                       0.0   \n",
       "plt1.run1 pat_89                                       0.0   \n",
       "...                                                    ...   \n",
       "plt49.run1 pat_3571                                    0.0   \n",
       "plt49.run1 pat_3572                                    0.0   \n",
       "\n",
       "                     rna_unass._chimeric_reads mapped_reads_over_2000000  \\\n",
       "rna project.run id                                                         \n",
       "plt1.run1 pat_88                           0.0                      True   \n",
       "plt1.run1 pat_89                           0.0                      True   \n",
       "...                                        ...                       ...   \n",
       "plt49.run1 pat_3571                        0.0                         0   \n",
       "plt49.run1 pat_3572                        0.0                         0   \n",
       "\n",
       "                     rna_DDX3Y  rna_XIST rna_gender rna_sample_id_original  \\\n",
       "rna project.run id                                                           \n",
       "plt1.run1 pat_88         844.0       8.0          M              Sample 49   \n",
       "plt1.run1 pat_89         670.0       4.0          M              Sample 50   \n",
       "...                        ...       ...        ...                    ...   \n",
       "plt49.run1 pat_3571        0.0       0.0          0                      0   \n",
       "plt49.run1 pat_3572        0.0       0.0          0                      0   \n",
       "\n",
       "                    rna_plate_id_original    filename rel_path  \n",
       "rna project.run id                                              \n",
       "plt1.run1 pat_88                  Plate 1  matrix.mtx        .  \n",
       "plt1.run1 pat_89                  Plate 1  matrix.mtx        .  \n",
       "...                                   ...         ...      ...  \n",
       "plt49.run1 pat_3571                     0  matrix.mtx        .  \n",
       "plt49.run1 pat_3572                     0  matrix.mtx        .  \n",
       "\n",
       "[3661 rows x 76 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl_gene</th>\n",
       "      <th>Ensembl_biotype</th>\n",
       "      <th>UniProt_proteins</th>\n",
       "      <th>UniProt_genes</th>\n",
       "      <th>UniProt_accessions</th>\n",
       "      <th>UniProt_names</th>\n",
       "      <th>UniProt_Ensembl_transcripts</th>\n",
       "      <th>UniProt_Ensembl_proteins</th>\n",
       "      <th>UniProt_Ensembl_gene_ids</th>\n",
       "      <th>Ensembl_chr</th>\n",
       "      <th>Ensembl_start</th>\n",
       "      <th>Ensembl_stop</th>\n",
       "      <th>Ensembl_strand</th>\n",
       "      <th>KEGG_PATHWAY_terms</th>\n",
       "      <th>KEGG_PATHWAY_ids</th>\n",
       "      <th>Gene_Ontology_terms</th>\n",
       "      <th>Gene_Ontology_ids</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ensembl gene id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000146555</th>\n",
       "      <td>SDK1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Protein sidekick-1</td>\n",
       "      <td>SDK1</td>\n",
       "      <td>Q7Z5N4; F8W6X9; A0A087WTQ6</td>\n",
       "      <td>SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN</td>\n",
       "      <td>ENST00000404826; ENST00000389531; ENST00000615806</td>\n",
       "      <td>ENSP00000385899; ENSP00000374182; ENSP00000478062</td>\n",
       "      <td>ENSG00000146555</td>\n",
       "      <td>chr7</td>\n",
       "      <td>3301252</td>\n",
       "      <td>4269000</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>eye development, neural retina development, bi...</td>\n",
       "      <td>GO:0001654, GO:0003407, GO:0005488, GO:0005515...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000283537</th>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000283537</td>\n",
       "      <td>chr7</td>\n",
       "      <td>143620943</td>\n",
       "      <td>143645675</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000270696</th>\n",
       "      <td>ENSG00000270696</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000270696</td>\n",
       "      <td>chr2</td>\n",
       "      <td>75660462</td>\n",
       "      <td>75662208</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000229906</th>\n",
       "      <td>SNRPGP11</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000229906</td>\n",
       "      <td>chr13</td>\n",
       "      <td>51496020</td>\n",
       "      <td>51496251</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57031 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Ensembl_gene                     Ensembl_biotype  \\\n",
       "Ensembl gene id                                                        \n",
       "ENSG00000146555             SDK1                      protein_coding   \n",
       "ENSG00000283537  ENSG00000283537  transcribed_unprocessed_pseudogene   \n",
       "...                          ...                                 ...   \n",
       "ENSG00000270696  ENSG00000270696                              lncRNA   \n",
       "ENSG00000229906         SNRPGP11                processed_pseudogene   \n",
       "\n",
       "                   UniProt_proteins UniProt_genes          UniProt_accessions  \\\n",
       "Ensembl gene id                                                                 \n",
       "ENSG00000146555  Protein sidekick-1          SDK1  Q7Z5N4; F8W6X9; A0A087WTQ6   \n",
       "ENSG00000283537                 NaN           NaN                         NaN   \n",
       "...                             ...           ...                         ...   \n",
       "ENSG00000270696                 NaN           NaN                         NaN   \n",
       "ENSG00000229906                 NaN           NaN                         NaN   \n",
       "\n",
       "                                              UniProt_names  \\\n",
       "Ensembl gene id                                               \n",
       "ENSG00000146555  SDK1_HUMAN; F8W6X9_HUMAN; A0A087WTQ6_HUMAN   \n",
       "ENSG00000283537                                         NaN   \n",
       "...                                                     ...   \n",
       "ENSG00000270696                                         NaN   \n",
       "ENSG00000229906                                         NaN   \n",
       "\n",
       "                                       UniProt_Ensembl_transcripts  \\\n",
       "Ensembl gene id                                                      \n",
       "ENSG00000146555  ENST00000404826; ENST00000389531; ENST00000615806   \n",
       "ENSG00000283537                                                NaN   \n",
       "...                                                            ...   \n",
       "ENSG00000270696                                                NaN   \n",
       "ENSG00000229906                                                NaN   \n",
       "\n",
       "                                          UniProt_Ensembl_proteins  \\\n",
       "Ensembl gene id                                                      \n",
       "ENSG00000146555  ENSP00000385899; ENSP00000374182; ENSP00000478062   \n",
       "ENSG00000283537                                                NaN   \n",
       "...                                                            ...   \n",
       "ENSG00000270696                                                NaN   \n",
       "ENSG00000229906                                                NaN   \n",
       "\n",
       "                UniProt_Ensembl_gene_ids Ensembl_chr  Ensembl_start  \\\n",
       "Ensembl gene id                                                       \n",
       "ENSG00000146555          ENSG00000146555        chr7        3301252   \n",
       "ENSG00000283537          ENSG00000283537        chr7      143620943   \n",
       "...                                  ...         ...            ...   \n",
       "ENSG00000270696          ENSG00000270696        chr2       75660462   \n",
       "ENSG00000229906          ENSG00000229906       chr13       51496020   \n",
       "\n",
       "                 Ensembl_stop Ensembl_strand KEGG_PATHWAY_terms  \\\n",
       "Ensembl gene id                                                   \n",
       "ENSG00000146555       4269000              +                NaN   \n",
       "ENSG00000283537     143645675              +                NaN   \n",
       "...                       ...            ...                ...   \n",
       "ENSG00000270696      75662208              +                NaN   \n",
       "ENSG00000229906      51496251              -                NaN   \n",
       "\n",
       "                KEGG_PATHWAY_ids  \\\n",
       "Ensembl gene id                    \n",
       "ENSG00000146555              NaN   \n",
       "ENSG00000283537              NaN   \n",
       "...                          ...   \n",
       "ENSG00000270696              NaN   \n",
       "ENSG00000229906              NaN   \n",
       "\n",
       "                                               Gene_Ontology_terms  \\\n",
       "Ensembl gene id                                                      \n",
       "ENSG00000146555  eye development, neural retina development, bi...   \n",
       "ENSG00000283537                                                NaN   \n",
       "...                                                            ...   \n",
       "ENSG00000270696                                                NaN   \n",
       "ENSG00000229906                                                NaN   \n",
       "\n",
       "                                                 Gene_Ontology_ids  \n",
       "Ensembl gene id                                                     \n",
       "ENSG00000146555  GO:0001654, GO:0003407, GO:0005488, GO:0005515...  \n",
       "ENSG00000283537                                                NaN  \n",
       "...                                                            ...  \n",
       "ENSG00000270696                                                NaN  \n",
       "ENSG00000229906                                                NaN  \n",
       "\n",
       "[57031 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(adata) # Übersicht\n",
    "with pd.option_context('display.max_rows', 5,'display.max_columns', None):\n",
    "    display(adata.obs)\n",
    "    display(adata.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] The adata object was saved to: anndata_1.h5ad\n"
     ]
    }
   ],
   "source": [
    "adata_output = \"anndata_1.h5ad\"\n",
    "utils.adata.save_h5ad(adata, adata_output) # speichern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "sctoolbox.settings.close_logfile()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sctoolbox",
   "language": "python",
   "name": "sctoolbox"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
